{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Adjust Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "#import libraries\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "#import data\n",
    "combinedF = os.path.join(\"C:/Users/jpcor/GoogleDrive/Data Science/Clauset's/PCR Analysis/UncleanedData/TestRaw/\", \"TESTredo_VanessaPCR_Reaction_Collection_4-9-21.csv\")  #\"TEST2A_VanessaPCR_Reaction_Collection_3-29-21.csv\"\n",
    "PCR_Data = pd.read_csv(combinedF)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>timestamp</th>\n",
       "      <th>user</th>\n",
       "      <th>email</th>\n",
       "      <th>species</th>\n",
       "      <th>primer1</th>\n",
       "      <th>primer2</th>\n",
       "      <th>amtprimer1</th>\n",
       "      <th>amtprimer2</th>\n",
       "      <th>polymerase_name</th>\n",
       "      <th>pcrid</th>\n",
       "      <th>...</th>\n",
       "      <th>template_type</th>\n",
       "      <th>dna_amt_units</th>\n",
       "      <th>wrong_bands</th>\n",
       "      <th>right_band</th>\n",
       "      <th>no_bands</th>\n",
       "      <th>band_size_units</th>\n",
       "      <th>customized_salts</th>\n",
       "      <th>buffer</th>\n",
       "      <th>dmso_amount</th>\n",
       "      <th>notes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>04/09/2021 14:31:50</td>\n",
       "      <td>Vanessa Lamae Bauer DuMont</td>\n",
       "      <td>vaba7447@colorado.edu</td>\n",
       "      <td>aotus azarae</td>\n",
       "      <td>CAGAAAAGTAACATCCTAGAC</td>\n",
       "      <td>GTTCCGTTGTGTCATCAGG</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.5</td>\n",
       "      <td>Q5 High-Fidelity DNA Polymerase</td>\n",
       "      <td>alison's</td>\n",
       "      <td>...</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>μg</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>bp</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>04/09/2021 14:35:03</td>\n",
       "      <td>Vanessa Lamae Bauer DuMont</td>\n",
       "      <td>vaba7447@colorado.edu</td>\n",
       "      <td>aotus azarae</td>\n",
       "      <td>TGCACCCACTGTCTTCC</td>\n",
       "      <td>AGGTCTTGCCGAAACAGATG</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.5</td>\n",
       "      <td>Q5 High-Fidelity DNA Polymerase</td>\n",
       "      <td>alison's</td>\n",
       "      <td>...</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>μg</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>bp</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>04/09/2021 14:48:50</td>\n",
       "      <td>Vanessa Lamae Bauer DuMont</td>\n",
       "      <td>vaba7447@colorado.edu</td>\n",
       "      <td>aotus azarae</td>\n",
       "      <td>GATGGGTGGCAAGTGGTC</td>\n",
       "      <td>AAGTTGACCAGTGCCGTTCC</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.5</td>\n",
       "      <td>Q5 High-Fidelity DNA Polymerase</td>\n",
       "      <td>alison's</td>\n",
       "      <td>...</td>\n",
       "      <td>Plasmid</td>\n",
       "      <td>μg</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>bp</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9 rows × 42 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             timestamp                        user                  email  \\\n",
       "0  04/09/2021 14:31:50  Vanessa Lamae Bauer DuMont  vaba7447@colorado.edu   \n",
       "1  04/09/2021 14:35:03  Vanessa Lamae Bauer DuMont  vaba7447@colorado.edu   \n",
       "2  04/09/2021 14:48:50  Vanessa Lamae Bauer DuMont  vaba7447@colorado.edu   \n",
       "3                  NaN                         NaN                    NaN   \n",
       "4                  NaN                         NaN                    NaN   \n",
       "5                  NaN                         NaN                    NaN   \n",
       "6                  NaN                         NaN                    NaN   \n",
       "7                  NaN                         NaN                    NaN   \n",
       "8                  NaN                         NaN                    NaN   \n",
       "\n",
       "        species                primer1               primer2  amtprimer1  \\\n",
       "0  aotus azarae  CAGAAAAGTAACATCCTAGAC   GTTCCGTTGTGTCATCAGG         0.5   \n",
       "1  aotus azarae      TGCACCCACTGTCTTCC  AGGTCTTGCCGAAACAGATG         0.5   \n",
       "2  aotus azarae     GATGGGTGGCAAGTGGTC  AAGTTGACCAGTGCCGTTCC         0.5   \n",
       "3           NaN                    NaN                   NaN         NaN   \n",
       "4           NaN                    NaN                   NaN         NaN   \n",
       "5           NaN                    NaN                   NaN         NaN   \n",
       "6           NaN                    NaN                   NaN         NaN   \n",
       "7           NaN                    NaN                   NaN         NaN   \n",
       "8           NaN                    NaN                   NaN         NaN   \n",
       "\n",
       "   amtprimer2                  polymerase_name     pcrid  ...  template_type  \\\n",
       "0         0.5  Q5 High-Fidelity DNA Polymerase  alison's  ...           cDNA   \n",
       "1         0.5  Q5 High-Fidelity DNA Polymerase  alison's  ...           cDNA   \n",
       "2         0.5  Q5 High-Fidelity DNA Polymerase  alison's  ...        Plasmid   \n",
       "3         NaN                              NaN       NaN  ...            NaN   \n",
       "4         NaN                              NaN       NaN  ...            NaN   \n",
       "5         NaN                              NaN       NaN  ...            NaN   \n",
       "6         NaN                              NaN       NaN  ...            NaN   \n",
       "7         NaN                              NaN       NaN  ...            NaN   \n",
       "8         NaN                              NaN       NaN  ...            NaN   \n",
       "\n",
       "   dna_amt_units  wrong_bands  right_band  no_bands  band_size_units  \\\n",
       "0             μg        False        True     False               bp   \n",
       "1             μg        False        True     False               bp   \n",
       "2             μg        False        True     False               bp   \n",
       "3            NaN          NaN         NaN       NaN              NaN   \n",
       "4            NaN          NaN         NaN       NaN              NaN   \n",
       "5            NaN          NaN         NaN       NaN              NaN   \n",
       "6            NaN          NaN         NaN       NaN              NaN   \n",
       "7            NaN          NaN         NaN       NaN              NaN   \n",
       "8            NaN          NaN         NaN       NaN              NaN   \n",
       "\n",
       "   customized_salts  buffer  dmso_amount  notes  \n",
       "0             False     NaN          NaN    NaN  \n",
       "1             False     NaN          NaN    NaN  \n",
       "2             False     NaN          NaN    NaN  \n",
       "3               NaN     NaN          NaN    NaN  \n",
       "4               NaN     NaN          NaN    NaN  \n",
       "5               NaN     NaN          NaN    NaN  \n",
       "6               NaN     NaN          NaN    NaN  \n",
       "7               NaN     NaN          NaN    NaN  \n",
       "8               NaN     NaN          NaN    NaN  \n",
       "\n",
       "[9 rows x 42 columns]"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PCR_Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "before dropping duplicates 9\n",
      "after dropping duplicates 4\n"
     ]
    }
   ],
   "source": [
    "#Drop duplicate data 10/26/20\n",
    "print('before dropping duplicates',len(PCR_Data))\n",
    "PCR_Data.drop_duplicates(keep='first',inplace=True)\n",
    "print('after dropping duplicates',len(PCR_Data))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#Fix polymerase name\n",
    "PCR_Data['polymerase_name']=PCR_Data['polymerase_name'].apply(lambda x: 'PhusionHF')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>timestamp</th>\n",
       "      <th>user</th>\n",
       "      <th>email</th>\n",
       "      <th>species</th>\n",
       "      <th>primer1</th>\n",
       "      <th>primer2</th>\n",
       "      <th>amtprimer1</th>\n",
       "      <th>amtprimer2</th>\n",
       "      <th>polymerase_name</th>\n",
       "      <th>pcrid</th>\n",
       "      <th>...</th>\n",
       "      <th>template_type</th>\n",
       "      <th>dna_amt_units</th>\n",
       "      <th>wrong_bands</th>\n",
       "      <th>right_band</th>\n",
       "      <th>no_bands</th>\n",
       "      <th>band_size_units</th>\n",
       "      <th>customized_salts</th>\n",
       "      <th>buffer</th>\n",
       "      <th>dmso_amount</th>\n",
       "      <th>notes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>04/09/2021 14:31:50</td>\n",
       "      <td>Vanessa Lamae Bauer DuMont</td>\n",
       "      <td>vaba7447@colorado.edu</td>\n",
       "      <td>aotus azarae</td>\n",
       "      <td>CAGAAAAGTAACATCCTAGAC</td>\n",
       "      <td>GTTCCGTTGTGTCATCAGG</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.5</td>\n",
       "      <td>Q5 High-Fidelity DNA Polymerase</td>\n",
       "      <td>alison's</td>\n",
       "      <td>...</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>μg</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>bp</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>04/09/2021 14:35:03</td>\n",
       "      <td>Vanessa Lamae Bauer DuMont</td>\n",
       "      <td>vaba7447@colorado.edu</td>\n",
       "      <td>aotus azarae</td>\n",
       "      <td>TGCACCCACTGTCTTCC</td>\n",
       "      <td>AGGTCTTGCCGAAACAGATG</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.5</td>\n",
       "      <td>Q5 High-Fidelity DNA Polymerase</td>\n",
       "      <td>alison's</td>\n",
       "      <td>...</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>μg</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>bp</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>04/09/2021 14:48:50</td>\n",
       "      <td>Vanessa Lamae Bauer DuMont</td>\n",
       "      <td>vaba7447@colorado.edu</td>\n",
       "      <td>aotus azarae</td>\n",
       "      <td>GATGGGTGGCAAGTGGTC</td>\n",
       "      <td>AAGTTGACCAGTGCCGTTCC</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.5</td>\n",
       "      <td>Q5 High-Fidelity DNA Polymerase</td>\n",
       "      <td>alison's</td>\n",
       "      <td>...</td>\n",
       "      <td>Plasmid</td>\n",
       "      <td>μg</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>bp</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4 rows × 42 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             timestamp                        user                  email  \\\n",
       "0  04/09/2021 14:31:50  Vanessa Lamae Bauer DuMont  vaba7447@colorado.edu   \n",
       "1  04/09/2021 14:35:03  Vanessa Lamae Bauer DuMont  vaba7447@colorado.edu   \n",
       "2  04/09/2021 14:48:50  Vanessa Lamae Bauer DuMont  vaba7447@colorado.edu   \n",
       "3                  NaN                         NaN                    NaN   \n",
       "\n",
       "        species                primer1               primer2  amtprimer1  \\\n",
       "0  aotus azarae  CAGAAAAGTAACATCCTAGAC   GTTCCGTTGTGTCATCAGG         0.5   \n",
       "1  aotus azarae      TGCACCCACTGTCTTCC  AGGTCTTGCCGAAACAGATG         0.5   \n",
       "2  aotus azarae     GATGGGTGGCAAGTGGTC  AAGTTGACCAGTGCCGTTCC         0.5   \n",
       "3           NaN                    NaN                   NaN         NaN   \n",
       "\n",
       "   amtprimer2                  polymerase_name     pcrid  ...  template_type  \\\n",
       "0         0.5  Q5 High-Fidelity DNA Polymerase  alison's  ...           cDNA   \n",
       "1         0.5  Q5 High-Fidelity DNA Polymerase  alison's  ...           cDNA   \n",
       "2         0.5  Q5 High-Fidelity DNA Polymerase  alison's  ...        Plasmid   \n",
       "3         NaN                              NaN       NaN  ...            NaN   \n",
       "\n",
       "   dna_amt_units  wrong_bands  right_band  no_bands  band_size_units  \\\n",
       "0             μg        False        True     False               bp   \n",
       "1             μg        False        True     False               bp   \n",
       "2             μg        False        True     False               bp   \n",
       "3            NaN          NaN         NaN       NaN              NaN   \n",
       "\n",
       "   customized_salts  buffer  dmso_amount  notes  \n",
       "0             False     NaN          NaN    NaN  \n",
       "1             False     NaN          NaN    NaN  \n",
       "2             False     NaN          NaN    NaN  \n",
       "3               NaN     NaN          NaN    NaN  \n",
       "\n",
       "[4 rows x 42 columns]"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PCR_Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Drop unnecessary columns\n",
    "remove=[\"timestamp\",\"user\",\"email\",\"species\",\"polymerase_catalog_number\",\"pcrid\",\n",
    "        \"template_type\",\"polymerase_brand\",\"dna_amt_units\",\"band_size_units\",'customized_salts']\n",
    "PCR_Data.drop(remove,axis=1,inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "PCR_Data=PCR_Data.head(len(PCR_Data)-1).copy(deep=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>primer1</th>\n",
       "      <th>primer2</th>\n",
       "      <th>amtprimer1</th>\n",
       "      <th>amtprimer2</th>\n",
       "      <th>polymerase_name</th>\n",
       "      <th>t1_min</th>\n",
       "      <th>temp1</th>\n",
       "      <th>cycles1</th>\n",
       "      <th>melting_t2_min</th>\n",
       "      <th>melting_temp2</th>\n",
       "      <th>...</th>\n",
       "      <th>melting_t2_sec</th>\n",
       "      <th>annealing_t2_sec</th>\n",
       "      <th>extension_t2_sec</th>\n",
       "      <th>t5_sec</th>\n",
       "      <th>wrong_bands</th>\n",
       "      <th>right_band</th>\n",
       "      <th>no_bands</th>\n",
       "      <th>buffer</th>\n",
       "      <th>dmso_amount</th>\n",
       "      <th>notes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CAGAAAAGTAACATCCTAGAC</td>\n",
       "      <td>GTTCCGTTGTGTCATCAGG</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.5</td>\n",
       "      <td>Q5 High-Fidelity DNA Polymerase</td>\n",
       "      <td>0.0</td>\n",
       "      <td>98.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>98.0</td>\n",
       "      <td>...</td>\n",
       "      <td>10.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TGCACCCACTGTCTTCC</td>\n",
       "      <td>AGGTCTTGCCGAAACAGATG</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.5</td>\n",
       "      <td>Q5 High-Fidelity DNA Polymerase</td>\n",
       "      <td>0.0</td>\n",
       "      <td>98.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>98.0</td>\n",
       "      <td>...</td>\n",
       "      <td>10.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GATGGGTGGCAAGTGGTC</td>\n",
       "      <td>AAGTTGACCAGTGCCGTTCC</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.5</td>\n",
       "      <td>Q5 High-Fidelity DNA Polymerase</td>\n",
       "      <td>0.0</td>\n",
       "      <td>98.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>98.0</td>\n",
       "      <td>...</td>\n",
       "      <td>10.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 31 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 primer1               primer2  amtprimer1  amtprimer2  \\\n",
       "0  CAGAAAAGTAACATCCTAGAC   GTTCCGTTGTGTCATCAGG         0.5         0.5   \n",
       "1      TGCACCCACTGTCTTCC  AGGTCTTGCCGAAACAGATG         0.5         0.5   \n",
       "2     GATGGGTGGCAAGTGGTC  AAGTTGACCAGTGCCGTTCC         0.5         0.5   \n",
       "\n",
       "                   polymerase_name  t1_min  temp1  cycles1  melting_t2_min  \\\n",
       "0  Q5 High-Fidelity DNA Polymerase     0.0   98.0      1.0             0.0   \n",
       "1  Q5 High-Fidelity DNA Polymerase     0.0   98.0      1.0             0.0   \n",
       "2  Q5 High-Fidelity DNA Polymerase     0.0   98.0      1.0             0.0   \n",
       "\n",
       "   melting_temp2  ...  melting_t2_sec  annealing_t2_sec  extension_t2_sec  \\\n",
       "0           98.0  ...            10.0              30.0               0.0   \n",
       "1           98.0  ...            10.0              30.0               0.0   \n",
       "2           98.0  ...            10.0              30.0               0.0   \n",
       "\n",
       "   t5_sec  wrong_bands  right_band  no_bands  buffer  dmso_amount  notes  \n",
       "0     0.0        False        True     False     NaN          NaN    NaN  \n",
       "1     0.0        False        True     False     NaN          NaN    NaN  \n",
       "2     0.0        False        True     False     NaN          NaN    NaN  \n",
       "\n",
       "[3 rows x 31 columns]"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PCR_Data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Set Outcomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Drop original output columns\n",
    "remove=[\"wrong_bands\",\"right_band\",\"no_bands\"]\n",
    "PCR_Data.drop(remove,axis=1,inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Data Cleaning"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "#rename incorrect column labels to correct column labels\n",
    "PCR_Data.rename(columns={'extension_t2_min.1':'extension_t2_sec','t5_sec':'t3_sec'},inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Convert time inputs to seconds\n",
    "PCR_Data[\"t1(sec)\"] = (PCR_Data[\"t1_min\"]*60) + PCR_Data[\"t1_sec\"]\n",
    "PCR_Data[\"melting_t2(sec)\"] = (PCR_Data[\"melting_t2_min\"]*60) + PCR_Data[\"melting_t2_sec\"]\n",
    "PCR_Data[\"annealing_t3(sec)\"] = (PCR_Data[\"annealing_t2_min\"]*60) + PCR_Data[\"annealing_t2_sec\"]\n",
    "PCR_Data[\"extension_t4(sec)\"] = (PCR_Data[\"extension_t2_min\"]*60) + PCR_Data[\"extension_t2_sec\"]\n",
    "PCR_Data[\"t5(sec)\"] = (PCR_Data[\"t3_min\"]*60) + PCR_Data[\"t3_sec\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "#drop old time columns\n",
    "remove=[\"t1_min\",\"t1_sec\",\"melting_t2_min\",\"melting_t2_sec\",\"annealing_t2_min\",\n",
    "        \"annealing_t2_sec\",\"extension_t2_min\",\"extension_t2_sec\",\"t3_min\",\"t3_sec\"]\n",
    "PCR_Data.drop(remove,axis=1,inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>primer1</th>\n",
       "      <th>primer2</th>\n",
       "      <th>amtprimer1</th>\n",
       "      <th>amtprimer2</th>\n",
       "      <th>polymerase_name</th>\n",
       "      <th>temp1</th>\n",
       "      <th>cycles1</th>\n",
       "      <th>melting_temp2</th>\n",
       "      <th>annealing_temp2</th>\n",
       "      <th>extension_temp2</th>\n",
       "      <th>...</th>\n",
       "      <th>band_size</th>\n",
       "      <th>dna_amt</th>\n",
       "      <th>buffer</th>\n",
       "      <th>dmso_amount</th>\n",
       "      <th>notes</th>\n",
       "      <th>t1(sec)</th>\n",
       "      <th>melting_t2(sec)</th>\n",
       "      <th>annealing_t3(sec)</th>\n",
       "      <th>extension_t4(sec)</th>\n",
       "      <th>t5(sec)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CAGAAAAGTAACATCCTAGAC</td>\n",
       "      <td>GTTCCGTTGTGTCATCAGG</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.5</td>\n",
       "      <td>Q5 High-Fidelity DNA Polymerase</td>\n",
       "      <td>98.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>98.0</td>\n",
       "      <td>58.0</td>\n",
       "      <td>72.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1840.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>30.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>60.0</td>\n",
       "      <td>120.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TGCACCCACTGTCTTCC</td>\n",
       "      <td>AGGTCTTGCCGAAACAGATG</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.5</td>\n",
       "      <td>Q5 High-Fidelity DNA Polymerase</td>\n",
       "      <td>98.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>98.0</td>\n",
       "      <td>66.0</td>\n",
       "      <td>72.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1423.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>30.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>60.0</td>\n",
       "      <td>120.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GATGGGTGGCAAGTGGTC</td>\n",
       "      <td>AAGTTGACCAGTGCCGTTCC</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.5</td>\n",
       "      <td>Q5 High-Fidelity DNA Polymerase</td>\n",
       "      <td>98.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>98.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>72.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2453.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>30.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>90.0</td>\n",
       "      <td>60.0</td>\n",
       "      <td>120.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 primer1               primer2  amtprimer1  amtprimer2  \\\n",
       "0  CAGAAAAGTAACATCCTAGAC   GTTCCGTTGTGTCATCAGG         0.5         0.5   \n",
       "1      TGCACCCACTGTCTTCC  AGGTCTTGCCGAAACAGATG         0.5         0.5   \n",
       "2     GATGGGTGGCAAGTGGTC  AAGTTGACCAGTGCCGTTCC         0.5         0.5   \n",
       "\n",
       "                   polymerase_name  temp1  cycles1  melting_temp2  \\\n",
       "0  Q5 High-Fidelity DNA Polymerase   98.0      1.0           98.0   \n",
       "1  Q5 High-Fidelity DNA Polymerase   98.0      1.0           98.0   \n",
       "2  Q5 High-Fidelity DNA Polymerase   98.0      1.0           98.0   \n",
       "\n",
       "   annealing_temp2  extension_temp2  ...  band_size  dna_amt  buffer  \\\n",
       "0             58.0             72.0  ...     1840.0      3.0     NaN   \n",
       "1             66.0             72.0  ...     1423.0      3.0     NaN   \n",
       "2             67.0             72.0  ...     2453.0      3.0     NaN   \n",
       "\n",
       "   dmso_amount  notes  t1(sec)  melting_t2(sec)  annealing_t3(sec)  \\\n",
       "0          NaN    NaN     30.0             10.0               30.0   \n",
       "1          NaN    NaN     30.0             10.0               30.0   \n",
       "2          NaN    NaN     30.0             10.0               90.0   \n",
       "\n",
       "   extension_t4(sec)  t5(sec)  \n",
       "0               60.0    120.0  \n",
       "1               60.0    120.0  \n",
       "2               60.0    120.0  \n",
       "\n",
       "[3 rows x 23 columns]"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PCR_Data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Feature Engineering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.SeqUtils import MeltingTemp as mt\n",
    "from Bio.SeqUtils import GC\n",
    "import primer3 as pr\n",
    "import melting as melt\n",
    "from NickFunctions import*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "def Tm(primer,polymerase,primerconc, option,primer2='NA'):\n",
    "    '''\n",
    "    \n",
    "    \n",
    "    Description:\n",
    "    ------------\n",
    "    Calculates primer melting temperature (Tm) based on a variety of options. Uilizes Buffer function to lookup salt concentrations for each function. Multiple options for testing.\n",
    "     NOTE: primer2 only necessary when calculating heterodimer\n",
    "    Options (default 3): \n",
    "    1 - Primer3 melting temperature\n",
    "    2 - Tm_NN BioSeqUtils\n",
    "    3 - melting (Virtually the same as IDT)\n",
    "    4 - Hairpin\n",
    "    5 - Homodimer\n",
    "    6 - Heterodimer need to input primer2\n",
    "    \n",
    "    Differences:\n",
    "\n",
    "    IDT unsure what NN method is used for calculation. Should use Owczarzy (2008) salt correction.\n",
    "\n",
    "    Primer3 uses santaLucia (1997) correction. Uses Owczarzy (2008) salt correction\n",
    "\n",
    "    BioSeqUtils uses santaLucia (2004) review correction. Uses Owczarzy (2008) salt correction\n",
    "    \n",
    "    Parameters:\n",
    "    ------------\n",
    "    primer, str: primer to calculate melting temperature for\n",
    "    polymerase, str: simplified polyerase name from \"ChangeName\" function. Used for lookup table of salt concentrations.\n",
    "    primerconc, float: from user input\n",
    "    option, int: melting temp prediction option\n",
    "    primer2, str: only needed for heterodimer calculation option.\n",
    "    \n",
    "    Return:\n",
    "    ------------\n",
    "    calculated melting temperature as float.\n",
    "    '''\n",
    "    \n",
    "    b=Buffer(polymerase)#salts in each polymerase buffer\n",
    "    Tms=[]\n",
    "    if (option==1):\n",
    "        return pr.calcTm(seq=primer,\n",
    "                                    mv_conc=b['MV_tot'],\n",
    "                                    dv_conc=b['DV_tot'],\n",
    "                                    dntp_conc=b['dNTPs'],\n",
    "                                    dna_conc=primerconc,\n",
    "                                    max_nn_length=60,\n",
    "                                    tm_method='santalucia',\n",
    "                                    salt_corrections_method=2)\n",
    "    elif (option==2):\n",
    "        return mt.Tm_NN(seq=primer,\n",
    "                                    nn_table=mt.DNA_NN4,\n",
    "                                    dnac1=primerconc,\n",
    "                                    dnac2=0,\n",
    "                                    Na=b['Na'], \n",
    "                                    K=b['K'],\n",
    "                                    Tris=b['Tris'],\n",
    "                                    Mg=b['Mg'],\n",
    "                                    dNTPs=b['dNTPs'],\n",
    "                                    saltcorr=7)\n",
    "    elif (option==3):\n",
    "        return melt.temp(primer,\n",
    "                                    DNA_c=primerconc,\n",
    "                                    Na_c=b['MV_tot'],\n",
    "                                    Mg_c=b['DV_tot'],\n",
    "                                    dNTPs_c=b['dNTPs'])\n",
    "    elif (option==4):\n",
    "        return pr.calcHairpin(seq=primer,\n",
    "                                    mv_conc=b['MV_tot'],\n",
    "                                    dv_conc=b['DV_tot'],\n",
    "                                    dntp_conc=b['dNTPs'],\n",
    "                                    dna_conc=primerconc,\n",
    "                                    temp_c=37,\n",
    "                                    max_loop=30,\n",
    "                                    output_structure=False).tm\n",
    "    elif (option==5):\n",
    "        return pr.calcHomodimer(seq=primer,\n",
    "                                    mv_conc=b['MV_tot'],\n",
    "                                    dv_conc=b['DV_tot'],\n",
    "                                    dntp_conc=b['dNTPs'],\n",
    "                                    dna_conc=primerconc,\n",
    "                                    temp_c=37,\n",
    "                                    max_loop=30,\n",
    "                                    output_structure=False).tm\n",
    "    elif (option==6):\n",
    "        return pr.calcHeterodimer(seq1=primer, seq2=primer2,\n",
    "                                    mv_conc=b['MV_tot'],\n",
    "                                    dv_conc=b['DV_tot'],\n",
    "                                    dntp_conc=b['dNTPs'],\n",
    "                                    dna_conc=primerconc,\n",
    "                                    temp_c=37,\n",
    "                                    max_loop=30,\n",
    "                                    output_structure=False).tm\n",
    "    else:\n",
    "        print(\"Not a valid option. Please input 1,2,3,4,5, or 6\")\n",
    "        return 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Change polymerase name to generalized form for Buffer dictionary\n",
    "PCR_Data[\"polymerase_name\"]=PCR_Data['polymerase_name'].apply(lambda x: ChangeName(x))\n",
    "#PCR_Data=PCR_Data.head(1).copy(deep=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>primer1</th>\n",
       "      <th>primer2</th>\n",
       "      <th>amtprimer1</th>\n",
       "      <th>amtprimer2</th>\n",
       "      <th>polymerase_name</th>\n",
       "      <th>temp1</th>\n",
       "      <th>cycles1</th>\n",
       "      <th>melting_temp2</th>\n",
       "      <th>annealing_temp2</th>\n",
       "      <th>extension_temp2</th>\n",
       "      <th>...</th>\n",
       "      <th>band_size</th>\n",
       "      <th>dna_amt</th>\n",
       "      <th>buffer</th>\n",
       "      <th>dmso_amount</th>\n",
       "      <th>notes</th>\n",
       "      <th>t1(sec)</th>\n",
       "      <th>melting_t2(sec)</th>\n",
       "      <th>annealing_t3(sec)</th>\n",
       "      <th>extension_t4(sec)</th>\n",
       "      <th>t5(sec)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CAGAAAAGTAACATCCTAGAC</td>\n",
       "      <td>GTTCCGTTGTGTCATCAGG</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.5</td>\n",
       "      <td>Q5</td>\n",
       "      <td>98.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>98.0</td>\n",
       "      <td>58.0</td>\n",
       "      <td>72.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1840.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>30.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>60.0</td>\n",
       "      <td>120.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TGCACCCACTGTCTTCC</td>\n",
       "      <td>AGGTCTTGCCGAAACAGATG</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.5</td>\n",
       "      <td>Q5</td>\n",
       "      <td>98.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>98.0</td>\n",
       "      <td>66.0</td>\n",
       "      <td>72.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1423.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>30.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>60.0</td>\n",
       "      <td>120.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GATGGGTGGCAAGTGGTC</td>\n",
       "      <td>AAGTTGACCAGTGCCGTTCC</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.5</td>\n",
       "      <td>Q5</td>\n",
       "      <td>98.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>98.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>72.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2453.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>30.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>90.0</td>\n",
       "      <td>60.0</td>\n",
       "      <td>120.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 primer1               primer2  amtprimer1  amtprimer2  \\\n",
       "0  CAGAAAAGTAACATCCTAGAC   GTTCCGTTGTGTCATCAGG         0.5         0.5   \n",
       "1      TGCACCCACTGTCTTCC  AGGTCTTGCCGAAACAGATG         0.5         0.5   \n",
       "2     GATGGGTGGCAAGTGGTC  AAGTTGACCAGTGCCGTTCC         0.5         0.5   \n",
       "\n",
       "  polymerase_name  temp1  cycles1  melting_temp2  annealing_temp2  \\\n",
       "0              Q5   98.0      1.0           98.0             58.0   \n",
       "1              Q5   98.0      1.0           98.0             66.0   \n",
       "2              Q5   98.0      1.0           98.0             67.0   \n",
       "\n",
       "   extension_temp2  ...  band_size  dna_amt  buffer  dmso_amount  notes  \\\n",
       "0             72.0  ...     1840.0      3.0     NaN          NaN    NaN   \n",
       "1             72.0  ...     1423.0      3.0     NaN          NaN    NaN   \n",
       "2             72.0  ...     2453.0      3.0     NaN          NaN    NaN   \n",
       "\n",
       "   t1(sec)  melting_t2(sec)  annealing_t3(sec)  extension_t4(sec)  t5(sec)  \n",
       "0     30.0             10.0               30.0               60.0    120.0  \n",
       "1     30.0             10.0               30.0               60.0    120.0  \n",
       "2     30.0             10.0               90.0               60.0    120.0  \n",
       "\n",
       "[3 rows x 23 columns]"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PCR_Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Strip blank spaces from primers\n",
    "PCR_Data[\"primer1\"] = PCR_Data[\"primer1\"].apply(lambda x: x.replace(' ',''))# remove all white space including middle of string\n",
    "PCR_Data[\"primer2\"] = PCR_Data[\"primer2\"].apply(lambda x: x.replace(' ',''))# remove all white space including middle of string\n",
    "#Ensure all primers are upper case\n",
    "PCR_Data[\"primer1\"] = PCR_Data[\"primer1\"].apply(lambda x: x.upper())\n",
    "PCR_Data[\"primer2\"] = PCR_Data[\"primer2\"].apply(lambda x: x.upper())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Add Primer GC content\n",
    "PCR_Data[\"primer1_GC\"] = PCR_Data[\"primer1\"].apply(lambda x: GC(x))\n",
    "PCR_Data[\"primer2_GC\"] = PCR_Data[\"primer2\"].apply(lambda x: GC(x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Add Primer Lengths\n",
    "PCR_Data[\"P1_len\"] = PCR_Data[\"primer1\"].apply(lambda x: len(x))\n",
    "PCR_Data[\"P2_len\"] = PCR_Data[\"primer2\"].apply(lambda x: len(x))\n",
    "#Remove primers with length greater than 60 bioinformatics algorithms do not work well when greater than 60\n",
    "longPs = PCR_Data[(PCR_Data[\"P1_len\"]>60) | (PCR_Data[\"P2_len\"]>60)].index \n",
    "PCR_Data.drop(longPs,inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>primer1</th>\n",
       "      <th>primer2</th>\n",
       "      <th>amtprimer1</th>\n",
       "      <th>amtprimer2</th>\n",
       "      <th>polymerase_name</th>\n",
       "      <th>temp1</th>\n",
       "      <th>cycles1</th>\n",
       "      <th>melting_temp2</th>\n",
       "      <th>annealing_temp2</th>\n",
       "      <th>extension_temp2</th>\n",
       "      <th>...</th>\n",
       "      <th>notes</th>\n",
       "      <th>t1(sec)</th>\n",
       "      <th>melting_t2(sec)</th>\n",
       "      <th>annealing_t3(sec)</th>\n",
       "      <th>extension_t4(sec)</th>\n",
       "      <th>t5(sec)</th>\n",
       "      <th>primer1_GC</th>\n",
       "      <th>primer2_GC</th>\n",
       "      <th>P1_len</th>\n",
       "      <th>P2_len</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CAGAAAAGTAACATCCTAGAC</td>\n",
       "      <td>GTTCCGTTGTGTCATCAGG</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.5</td>\n",
       "      <td>Q5</td>\n",
       "      <td>98.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>98.0</td>\n",
       "      <td>58.0</td>\n",
       "      <td>72.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>30.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>60.0</td>\n",
       "      <td>120.0</td>\n",
       "      <td>38.095238</td>\n",
       "      <td>52.631579</td>\n",
       "      <td>21</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TGCACCCACTGTCTTCC</td>\n",
       "      <td>AGGTCTTGCCGAAACAGATG</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.5</td>\n",
       "      <td>Q5</td>\n",
       "      <td>98.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>98.0</td>\n",
       "      <td>66.0</td>\n",
       "      <td>72.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>30.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>60.0</td>\n",
       "      <td>120.0</td>\n",
       "      <td>58.823529</td>\n",
       "      <td>50.000000</td>\n",
       "      <td>17</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GATGGGTGGCAAGTGGTC</td>\n",
       "      <td>AAGTTGACCAGTGCCGTTCC</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.5</td>\n",
       "      <td>Q5</td>\n",
       "      <td>98.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>98.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>72.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>30.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>90.0</td>\n",
       "      <td>60.0</td>\n",
       "      <td>120.0</td>\n",
       "      <td>61.111111</td>\n",
       "      <td>55.000000</td>\n",
       "      <td>18</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 primer1               primer2  amtprimer1  amtprimer2  \\\n",
       "0  CAGAAAAGTAACATCCTAGAC   GTTCCGTTGTGTCATCAGG         0.5         0.5   \n",
       "1      TGCACCCACTGTCTTCC  AGGTCTTGCCGAAACAGATG         0.5         0.5   \n",
       "2     GATGGGTGGCAAGTGGTC  AAGTTGACCAGTGCCGTTCC         0.5         0.5   \n",
       "\n",
       "  polymerase_name  temp1  cycles1  melting_temp2  annealing_temp2  \\\n",
       "0              Q5   98.0      1.0           98.0             58.0   \n",
       "1              Q5   98.0      1.0           98.0             66.0   \n",
       "2              Q5   98.0      1.0           98.0             67.0   \n",
       "\n",
       "   extension_temp2  ...  notes  t1(sec)  melting_t2(sec)  annealing_t3(sec)  \\\n",
       "0             72.0  ...    NaN     30.0             10.0               30.0   \n",
       "1             72.0  ...    NaN     30.0             10.0               30.0   \n",
       "2             72.0  ...    NaN     30.0             10.0               90.0   \n",
       "\n",
       "   extension_t4(sec)  t5(sec)  primer1_GC  primer2_GC  P1_len  P2_len  \n",
       "0               60.0    120.0   38.095238   52.631579      21      19  \n",
       "1               60.0    120.0   58.823529   50.000000      17      20  \n",
       "2               60.0    120.0   61.111111   55.000000      18      20  \n",
       "\n",
       "[3 rows x 27 columns]"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PCR_Data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "def tep(x):\n",
    "    return 'PhusionHF'\n",
    "PCR_Data['polymerase_name']=PCR_Data['polymerase_name'].apply(lambda x: tep(x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Add Primer melting temps\n",
    "#Primer 3 no correction original model\n",
    "PCR_Data[\"Tm_p1_p3orig\"]=PCR_Data.apply(lambda x: pr.calcTm(x.primer1),axis=1)\n",
    "PCR_Data[\"Tm_p2_p3orig\"]=PCR_Data.apply(lambda x: pr.calcTm(x.primer2),axis=1)\n",
    "#Primer3\n",
    "PCR_Data[\"Tm_p1_p3\"]=PCR_Data.apply(lambda x: Tm(primer=x.primer1,polymerase=x.polymerase_name,primerconc=(x.amtprimer1+x.amtprimer2),option=1),axis=1)\n",
    "PCR_Data[\"Tm_p2_p3\"]=PCR_Data.apply(lambda x: Tm(primer=x.primer2,polymerase=x.polymerase_name,primerconc=(x.amtprimer1+x.amtprimer2),option=1),axis=1)\n",
    "#BioseqUtils\n",
    "PCR_Data[\"Tm_p1_BSU\"]=PCR_Data.apply(lambda x: Tm(primer=x.primer1,polymerase=x.polymerase_name,primerconc=(x.amtprimer1+x.amtprimer2),option=2),axis=1)\n",
    "PCR_Data[\"Tm_p2_BSU\"]=PCR_Data.apply(lambda x: Tm(primer=x.primer2,polymerase=x.polymerase_name,primerconc=(x.amtprimer1+x.amtprimer2),option=2),axis=1)\n",
    "#Melting (IDT analog)\n",
    "PCR_Data[\"Tm_p1_IDT\"]=PCR_Data.apply(lambda x: Tm(primer=x.primer1,polymerase=x.polymerase_name,primerconc=(x.amtprimer1+x.amtprimer2),option=3),axis=1)\n",
    "PCR_Data[\"Tm_p2_IDT\"]=PCR_Data.apply(lambda x: Tm(primer=x.primer2,polymerase=x.polymerase_name,primerconc=(x.amtprimer1+x.amtprimer2),option=3),axis=1)\n",
    "#Average\n",
    "#PCR_Data[\"Tm_p1_Avg\"]=(PCR_Data[\"Tm_p1_p3\"]+PCR_Data[\"Tm_p1_BSU\"]+PCR_Data[\"Tm_p1_IDT\"])/3\n",
    "#PCR_Data[\"Tm_p2_Avg\"]=(PCR_Data[\"Tm_p2_p3\"]+PCR_Data[\"Tm_p2_BSU\"]+PCR_Data[\"Tm_p2_IDT\"])/3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "#difference in primer melting temps\n",
    "PCR_Data[\"Tm_dif_IDT\"]=abs(PCR_Data[\"Tm_p1_IDT\"]-PCR_Data[\"Tm_p2_IDT\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Add if hairpin structure is predicted\n",
    "PCR_Data[\"HP_P1\"] = PCR_Data.apply(lambda x: Tm(primer=x.primer1,polymerase=x.polymerase_name,primerconc=(x.amtprimer1+x.amtprimer2),option=4),axis=1)\n",
    "PCR_Data[\"HP_P2\"] = PCR_Data.apply(lambda x: Tm(primer=x.primer2,polymerase=x.polymerase_name,primerconc=(x.amtprimer1+x.amtprimer2),option=4),axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Combine hairpins to one feature that has the highest hairpin temp\n",
    "#if first statement is true return 2nd value first statement is false return third value.\n",
    "PCR_Data[\"HPin_Tm_Max\"] = np.where(PCR_Data[\"HP_P1\"]>PCR_Data[\"HP_P2\"],PCR_Data[\"HP_P1\"],PCR_Data[\"HP_P2\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3.0"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(PCR_Data.size)/len(PCR_Data.columns)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Note**\n",
    "\n",
    "May want to test how setting any negative values to 0 affects performance. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Calculate Heterodimer\n",
    "PCR_Data[\"Heterodimer_tm\"] = PCR_Data.apply(lambda x : Tm(primer=x.primer1,primer2=x.primer2,polymerase=x.polymerase_name,primerconc=(x.amtprimer1+x.amtprimer2),option=6),axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Calculate Homodimer\n",
    "PCR_Data[\"Homodimer_tm1\"] = PCR_Data.apply(lambda x : Tm(primer=x.primer1,polymerase=x.polymerase_name,primerconc=(x.amtprimer1+x.amtprimer2),option=5),axis=1)\n",
    "PCR_Data[\"Homodimer_tm2\"] = PCR_Data.apply(lambda x : Tm(primer=x.primer2,polymerase=x.polymerase_name,primerconc=(x.amtprimer1+x.amtprimer2),option=5),axis=1)\n",
    "PCR_Data[\"Homodimer_maxTm\"] = np.where(PCR_Data[\"Homodimer_tm1\"]>PCR_Data[\"Homodimer_tm2\"],PCR_Data[\"Homodimer_tm1\"],PCR_Data[\"Homodimer_tm2\"])\n",
    "PCR_Data[\"Homodimer_minTm\"] = np.where(PCR_Data[\"Homodimer_tm1\"]<PCR_Data[\"Homodimer_tm2\"],PCR_Data[\"Homodimer_tm1\"],PCR_Data[\"Homodimer_tm2\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "def changeTm(Tm):\n",
    "    '''\n",
    "    Tms of less than or equal to zero mean there was no significant structure. To help the RF make these temps all 0\n",
    "    '''\n",
    "    if Tm <0:\n",
    "        return 0\n",
    "    else:\n",
    "        return Tm\n",
    "PCR_Data[\"Homodimer_maxTm\"]=PCR_Data[\"Homodimer_maxTm\"].apply(lambda x: changeTm(x))\n",
    "PCR_Data[\"Heterodimer_tm\"]=PCR_Data[\"Heterodimer_tm\"].apply(lambda x: changeTm(x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [],
   "source": [
    "PCR_Data['MaxTm_IDT']=np.where(PCR_Data[\"Tm_p1_IDT\"]>PCR_Data[\"Tm_p2_IDT\"],PCR_Data[\"Tm_p1_IDT\"],PCR_Data[\"Tm_p2_IDT\"])\n",
    "PCR_Data['MinTm_IDT']=np.where(PCR_Data[\"Tm_p1_IDT\"]<PCR_Data[\"Tm_p2_IDT\"],PCR_Data[\"Tm_p1_IDT\"],PCR_Data[\"Tm_p2_IDT\"])\n",
    "PCR_Data['MaxGC']=np.where(PCR_Data[\"primer1_GC\"]>PCR_Data[\"primer2_GC\"],PCR_Data[\"primer1_GC\"],PCR_Data[\"primer2_GC\"])\n",
    "PCR_Data['MinGC']=np.where(PCR_Data[\"primer1_GC\"]<PCR_Data[\"primer2_GC\"],PCR_Data[\"primer1_GC\"],PCR_Data[\"primer2_GC\"])\n",
    "PCR_Data['MaxPlen']=np.where(PCR_Data[\"P1_len\"]>PCR_Data[\"P2_len\"],PCR_Data[\"P1_len\"],PCR_Data[\"P2_len\"])\n",
    "PCR_Data['MinPlen']=np.where(PCR_Data[\"P1_len\"]<PCR_Data[\"P2_len\"],PCR_Data[\"P1_len\"],PCR_Data[\"P2_len\"])\n",
    "PCR_Data['TotPrimer']=PCR_Data['amtprimer1']+PCR_Data['amtprimer2']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**GC Clamp:**\n",
    "\n",
    "According to Top Tip Bio (TTP) (https://toptipbio.com/gc-clamp-pcr/) GC Clamp is defined by a G or C in the last 5 nucleotides on the 3' end of the primer.\n",
    "\n",
    "Strength table for NN model taken from Khandelwahl 2010"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "class GC_Clamp_features(object):\n",
    "    '''\n",
    "    Description:\n",
    "    ------------\n",
    "    Class obtains features from GC clamp such as strength and num_GC \n",
    "    NN model strength table taken from \n",
    "    Khandelwal G, Bhyravabhotla J. A Phenomenological Model for Predicting Melting Temperatures of DNA Sequences. PLOS ONE. 2010;5: e12433. doi:10.1371/journal.pone.0012433\n",
    "    '''\n",
    "    #initialize class variables\n",
    "    clamp = \"\"\n",
    "    num_GC = 0\n",
    "    Strength={'GC': 13,'CC': 11,'GG': 11,'CG': 10,'AC': 10,'TC': 8,'AG': 8,'TG': 7,'GT': 10,'CT': 8,'GA': 8,'CA': 7,'AT': 7,'TT': 5,'AA': 5,'TA': 4}\n",
    "    Score=0\n",
    "    #GC_Clamp class constructor\n",
    "    def __init__(self,primer):\n",
    "        self.clamp = primer[-5:]\n",
    "        for nt in range(len(self.clamp)-1):#inrement by 1 nucleotide at a time\n",
    "            NN=self.clamp[nt]+self.clamp[nt+1]#grab dinucleotides\n",
    "            self.Score+=self.Strength[NN]#GC clamp strength score\n",
    "        for nt in range(len(self.clamp)):\n",
    "            if self.clamp[nt]==\"G\" or self.clamp[nt]==\"C\":\n",
    "                self.num_GC+=1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TAGCA\n",
      "2\n",
      "32\n"
     ]
    }
   ],
   "source": [
    "clamp = GC_Clamp_features(\"ATACGTACGATAGCA\")\n",
    "print(clamp.clamp)\n",
    "print(clamp.num_GC)\n",
    "print(clamp.Score)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Create GC_Clamp_features Score\n",
    "PCR_Data[\"GC_Clamp1\"] = PCR_Data[\"primer1\"].apply(lambda x: GC_Clamp_features(x).Score)\n",
    "PCR_Data[\"GC_Clamp2\"] = PCR_Data[\"primer2\"].apply(lambda x: GC_Clamp_features(x).Score)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finalize GC_Clamp features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Take the lowest GC in clamp and positional weight for p1 and p2\n",
    "PCR_Data[\"GC_ClampMin\"] = np.where(PCR_Data[\"GC_Clamp1\"]<PCR_Data[\"GC_Clamp2\"],PCR_Data[\"GC_Clamp1\"],PCR_Data[\"GC_Clamp2\"])\n",
    "PCR_Data[\"GC_ClampMax\"] = np.where(PCR_Data[\"GC_Clamp1\"]>PCR_Data[\"GC_Clamp2\"],PCR_Data[\"GC_Clamp1\"],PCR_Data[\"GC_Clamp2\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>temp1</th>\n",
       "      <th>melting_temp2</th>\n",
       "      <th>annealing_temp2</th>\n",
       "      <th>extension_temp2</th>\n",
       "      <th>cycles2</th>\n",
       "      <th>temp3</th>\n",
       "      <th>band_size</th>\n",
       "      <th>dna_amt</th>\n",
       "      <th>buffer</th>\n",
       "      <th>dmso_amount</th>\n",
       "      <th>...</th>\n",
       "      <th>Homodimer_minTm</th>\n",
       "      <th>MaxTm_IDT</th>\n",
       "      <th>MinTm_IDT</th>\n",
       "      <th>MaxGC</th>\n",
       "      <th>MinGC</th>\n",
       "      <th>MaxPlen</th>\n",
       "      <th>MinPlen</th>\n",
       "      <th>TotPrimer</th>\n",
       "      <th>GC_ClampMin</th>\n",
       "      <th>GC_ClampMax</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>98.0</td>\n",
       "      <td>98.0</td>\n",
       "      <td>58.0</td>\n",
       "      <td>72.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>72.0</td>\n",
       "      <td>1840.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>-55.773368</td>\n",
       "      <td>51.387895</td>\n",
       "      <td>47.132343</td>\n",
       "      <td>52.631579</td>\n",
       "      <td>38.095238</td>\n",
       "      <td>21</td>\n",
       "      <td>19</td>\n",
       "      <td>1.0</td>\n",
       "      <td>30</td>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>98.0</td>\n",
       "      <td>98.0</td>\n",
       "      <td>66.0</td>\n",
       "      <td>72.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>72.0</td>\n",
       "      <td>1423.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>-66.357220</td>\n",
       "      <td>53.493671</td>\n",
       "      <td>51.075794</td>\n",
       "      <td>58.823529</td>\n",
       "      <td>50.000000</td>\n",
       "      <td>20</td>\n",
       "      <td>17</td>\n",
       "      <td>1.0</td>\n",
       "      <td>30</td>\n",
       "      <td>32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>98.0</td>\n",
       "      <td>98.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>72.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>72.0</td>\n",
       "      <td>2453.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>-45.030107</td>\n",
       "      <td>55.996572</td>\n",
       "      <td>52.940547</td>\n",
       "      <td>61.111111</td>\n",
       "      <td>55.000000</td>\n",
       "      <td>20</td>\n",
       "      <td>18</td>\n",
       "      <td>1.0</td>\n",
       "      <td>34</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 36 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   temp1  melting_temp2  annealing_temp2  extension_temp2  cycles2  temp3  \\\n",
       "0   98.0           98.0             58.0             72.0     35.0   72.0   \n",
       "1   98.0           98.0             66.0             72.0     35.0   72.0   \n",
       "2   98.0           98.0             67.0             72.0     35.0   72.0   \n",
       "\n",
       "   band_size  dna_amt  buffer  dmso_amount  ...  Homodimer_minTm  MaxTm_IDT  \\\n",
       "0     1840.0      3.0     NaN          NaN  ...       -55.773368  51.387895   \n",
       "1     1423.0      3.0     NaN          NaN  ...       -66.357220  53.493671   \n",
       "2     2453.0      3.0     NaN          NaN  ...       -45.030107  55.996572   \n",
       "\n",
       "   MinTm_IDT      MaxGC      MinGC  MaxPlen  MinPlen  TotPrimer  GC_ClampMin  \\\n",
       "0  47.132343  52.631579  38.095238       21       19        1.0           30   \n",
       "1  51.075794  58.823529  50.000000       20       17        1.0           30   \n",
       "2  52.940547  61.111111  55.000000       20       18        1.0           34   \n",
       "\n",
       "   GC_ClampMax  \n",
       "0           34  \n",
       "1           32  \n",
       "2           36  \n",
       "\n",
       "[3 rows x 36 columns]"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Drop GC_CLamp columns used for analysis and primers because we are done pulling data from them\n",
    "remove=[\"GC_Clamp1\",\"GC_Clamp2\",\"primer1\",\"primer2\",\"polymerase_name\",\"cycles1\",\"cycles3\",\"HP_P1\",\"HP_P2\",\"Homodimer_tm1\",\"Homodimer_tm2\",\"Tm_p1_IDT\",\"Tm_p2_IDT\",\"primer1_GC\",\"primer2_GC\",\"P1_len\",\"P2_len\",'amtprimer1','amtprimer2']\n",
    "PCR_Data.drop(remove,axis=1,inplace=True)\n",
    "PCR_Data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>temp1</th>\n",
       "      <th>melting_temp2</th>\n",
       "      <th>annealing_temp2</th>\n",
       "      <th>extension_temp2</th>\n",
       "      <th>cycles2</th>\n",
       "      <th>temp3</th>\n",
       "      <th>band_size</th>\n",
       "      <th>dna_amt</th>\n",
       "      <th>buffer</th>\n",
       "      <th>dmso_amount</th>\n",
       "      <th>...</th>\n",
       "      <th>Homodimer_minTm</th>\n",
       "      <th>MaxTm_IDT</th>\n",
       "      <th>MinTm_IDT</th>\n",
       "      <th>MaxGC</th>\n",
       "      <th>MinGC</th>\n",
       "      <th>MaxPlen</th>\n",
       "      <th>MinPlen</th>\n",
       "      <th>TotPrimer</th>\n",
       "      <th>GC_ClampMin</th>\n",
       "      <th>GC_ClampMax</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>98.0</td>\n",
       "      <td>98.0</td>\n",
       "      <td>58.0</td>\n",
       "      <td>72.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>72.0</td>\n",
       "      <td>1840.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>-55.773368</td>\n",
       "      <td>51.387895</td>\n",
       "      <td>47.132343</td>\n",
       "      <td>52.631579</td>\n",
       "      <td>38.095238</td>\n",
       "      <td>21</td>\n",
       "      <td>19</td>\n",
       "      <td>1.0</td>\n",
       "      <td>30</td>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>98.0</td>\n",
       "      <td>98.0</td>\n",
       "      <td>66.0</td>\n",
       "      <td>72.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>72.0</td>\n",
       "      <td>1423.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>-66.357220</td>\n",
       "      <td>53.493671</td>\n",
       "      <td>51.075794</td>\n",
       "      <td>58.823529</td>\n",
       "      <td>50.000000</td>\n",
       "      <td>20</td>\n",
       "      <td>17</td>\n",
       "      <td>1.0</td>\n",
       "      <td>30</td>\n",
       "      <td>32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>98.0</td>\n",
       "      <td>98.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>72.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>72.0</td>\n",
       "      <td>2453.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>-45.030107</td>\n",
       "      <td>55.996572</td>\n",
       "      <td>52.940547</td>\n",
       "      <td>61.111111</td>\n",
       "      <td>55.000000</td>\n",
       "      <td>20</td>\n",
       "      <td>18</td>\n",
       "      <td>1.0</td>\n",
       "      <td>34</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 36 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   temp1  melting_temp2  annealing_temp2  extension_temp2  cycles2  temp3  \\\n",
       "0   98.0           98.0             58.0             72.0     35.0   72.0   \n",
       "1   98.0           98.0             66.0             72.0     35.0   72.0   \n",
       "2   98.0           98.0             67.0             72.0     35.0   72.0   \n",
       "\n",
       "   band_size  dna_amt  buffer  dmso_amount  ...  Homodimer_minTm  MaxTm_IDT  \\\n",
       "0     1840.0      3.0     NaN          NaN  ...       -55.773368  51.387895   \n",
       "1     1423.0      3.0     NaN          NaN  ...       -66.357220  53.493671   \n",
       "2     2453.0      3.0     NaN          NaN  ...       -45.030107  55.996572   \n",
       "\n",
       "   MinTm_IDT      MaxGC      MinGC  MaxPlen  MinPlen  TotPrimer  GC_ClampMin  \\\n",
       "0  47.132343  52.631579  38.095238       21       19        1.0           30   \n",
       "1  51.075794  58.823529  50.000000       20       17        1.0           30   \n",
       "2  52.940547  61.111111  55.000000       20       18        1.0           34   \n",
       "\n",
       "   GC_ClampMax  \n",
       "0           34  \n",
       "1           32  \n",
       "2           36  \n",
       "\n",
       "[3 rows x 36 columns]"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PCR_Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3.0"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PCR_Data['dna_amt'].min()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Move outcome to the first position\n",
    "PCR_Data=PCR_Data[['temp1', 'melting_temp2',\n",
    "       'annealing_temp2', 'extension_temp2', 'cycles2', 'temp3',\n",
    "       'band_size', 'dna_amt','t1(sec)', 'melting_t2(sec)',\n",
    "       'annealing_t3(sec)', 'extension_t4(sec)', 't5(sec)', 'Tm_p1_p3orig', 'Tm_p2_p3orig', 'Tm_p1_p3',\n",
    "       'Tm_p2_p3', 'Tm_p1_BSU', 'Tm_p2_BSU', 'Tm_dif_IDT',\n",
    "       'HPin_Tm_Max', 'Heterodimer_tm', 'Homodimer_maxTm', 'GC_ClampMin',\n",
    "       'GC_ClampMax', 'MaxTm_IDT','MinTm_IDT','MaxGC','MinGC','MaxPlen','MinPlen','TotPrimer']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Current date\n",
      "2021-04-12\n"
     ]
    }
   ],
   "source": [
    "import datetime\n",
    "now = datetime.datetime.now()\n",
    "print(\"Current date\")\n",
    "now = str(now)\n",
    "date = now.split(\" \")\n",
    "print(date[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Vanessa_Dumont_redo41921_2021-04-12_myeng110120_predictoutcome.csv\n"
     ]
    }
   ],
   "source": [
    "engfname = \"Vanessa_Dumont_redo41921_\"+date[0]+\"_myeng110120_predictoutcome.csv\"#0706 noticed error in primer feat eng\n",
    "print(engfname)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [],
   "source": [
    "cleanedF = os.path.join(\"C:/Users/jpcor/GoogleDrive/Data Science/Clauset's/PCR Analysis/CleanedData/TestData/\", engfname)  \n",
    "PCR_Data.to_csv(cleanedF,index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Drop HP_P1 and HP_P2\n",
    "remove=[\"HP_P1\",\"HP_P2\"]\n",
    "PCR_Data.drop(remove,axis=1,inplace=True)\n",
    "PCR_Data.head()\n",
    "remove=[\"Homodimer_tm1\",\"Homodimer_tm2\"]\n",
    "PCR_Data.drop(remove,axis=1,inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>temp1</th>\n",
       "      <th>melting_temp2</th>\n",
       "      <th>annealing_temp2</th>\n",
       "      <th>extension_temp2</th>\n",
       "      <th>cycles2</th>\n",
       "      <th>temp3</th>\n",
       "      <th>band_size</th>\n",
       "      <th>dna_amt</th>\n",
       "      <th>t1(sec)</th>\n",
       "      <th>melting_t2(sec)</th>\n",
       "      <th>...</th>\n",
       "      <th>Homodimer_maxTm</th>\n",
       "      <th>GC_ClampMin</th>\n",
       "      <th>GC_ClampMax</th>\n",
       "      <th>MaxTm_IDT</th>\n",
       "      <th>MinTm_IDT</th>\n",
       "      <th>MaxGC</th>\n",
       "      <th>MinGC</th>\n",
       "      <th>MaxPlen</th>\n",
       "      <th>MinPlen</th>\n",
       "      <th>TotPrimer</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>98.0</td>\n",
       "      <td>98.0</td>\n",
       "      <td>58.0</td>\n",
       "      <td>72.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>72.0</td>\n",
       "      <td>1840.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>30</td>\n",
       "      <td>34</td>\n",
       "      <td>50.109406</td>\n",
       "      <td>45.962008</td>\n",
       "      <td>52.631579</td>\n",
       "      <td>38.095238</td>\n",
       "      <td>21</td>\n",
       "      <td>19</td>\n",
       "      <td>0.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>98.0</td>\n",
       "      <td>98.0</td>\n",
       "      <td>66.0</td>\n",
       "      <td>72.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>72.0</td>\n",
       "      <td>1300.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>33</td>\n",
       "      <td>41</td>\n",
       "      <td>52.208138</td>\n",
       "      <td>49.222177</td>\n",
       "      <td>75.000000</td>\n",
       "      <td>62.500000</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>0.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>98.0</td>\n",
       "      <td>98.0</td>\n",
       "      <td>65.0</td>\n",
       "      <td>72.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>72.0</td>\n",
       "      <td>3000.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>32</td>\n",
       "      <td>36</td>\n",
       "      <td>54.292811</td>\n",
       "      <td>51.435699</td>\n",
       "      <td>50.000000</td>\n",
       "      <td>43.478261</td>\n",
       "      <td>23</td>\n",
       "      <td>20</td>\n",
       "      <td>0.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>98.0</td>\n",
       "      <td>98.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>72.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>72.0</td>\n",
       "      <td>3000.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>34</td>\n",
       "      <td>36</td>\n",
       "      <td>54.743410</td>\n",
       "      <td>51.564502</td>\n",
       "      <td>61.111111</td>\n",
       "      <td>55.000000</td>\n",
       "      <td>20</td>\n",
       "      <td>18</td>\n",
       "      <td>0.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>98.0</td>\n",
       "      <td>98.0</td>\n",
       "      <td>66.0</td>\n",
       "      <td>72.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>72.0</td>\n",
       "      <td>8300.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>30</td>\n",
       "      <td>31</td>\n",
       "      <td>54.277472</td>\n",
       "      <td>52.496881</td>\n",
       "      <td>60.000000</td>\n",
       "      <td>47.619048</td>\n",
       "      <td>21</td>\n",
       "      <td>20</td>\n",
       "      <td>0.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>98.0</td>\n",
       "      <td>98.0</td>\n",
       "      <td>66.0</td>\n",
       "      <td>72.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>72.0</td>\n",
       "      <td>1423.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>30</td>\n",
       "      <td>32</td>\n",
       "      <td>52.246785</td>\n",
       "      <td>49.615785</td>\n",
       "      <td>58.823529</td>\n",
       "      <td>50.000000</td>\n",
       "      <td>20</td>\n",
       "      <td>17</td>\n",
       "      <td>0.4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6 rows × 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   temp1  melting_temp2  annealing_temp2  extension_temp2  cycles2  temp3  \\\n",
       "0   98.0           98.0             58.0             72.0     35.0   72.0   \n",
       "1   98.0           98.0             66.0             72.0     35.0   72.0   \n",
       "2   98.0           98.0             65.0             72.0     35.0   72.0   \n",
       "3   98.0           98.0             67.0             72.0     35.0   72.0   \n",
       "4   98.0           98.0             66.0             72.0     35.0   72.0   \n",
       "5   98.0           98.0             66.0             72.0     35.0   72.0   \n",
       "\n",
       "   band_size  dna_amt  t1(sec)  melting_t2(sec)  ...  Homodimer_maxTm  \\\n",
       "0     1840.0      3.0     30.0             10.0  ...                0   \n",
       "1     1300.0      3.0     30.0             10.0  ...                0   \n",
       "2     3000.0      3.0     30.0             10.0  ...                0   \n",
       "3     3000.0      3.0     30.0             10.0  ...                0   \n",
       "4     8300.0      3.0     30.0             10.0  ...                0   \n",
       "5     1423.0      3.0     30.0             10.0  ...                0   \n",
       "\n",
       "   GC_ClampMin  GC_ClampMax  MaxTm_IDT  MinTm_IDT      MaxGC      MinGC  \\\n",
       "0           30           34  50.109406  45.962008  52.631579  38.095238   \n",
       "1           33           41  52.208138  49.222177  75.000000  62.500000   \n",
       "2           32           36  54.292811  51.435699  50.000000  43.478261   \n",
       "3           34           36  54.743410  51.564502  61.111111  55.000000   \n",
       "4           30           31  54.277472  52.496881  60.000000  47.619048   \n",
       "5           30           32  52.246785  49.615785  58.823529  50.000000   \n",
       "\n",
       "   MaxPlen  MinPlen  TotPrimer  \n",
       "0       21       19        0.4  \n",
       "1       16       16        0.4  \n",
       "2       23       20        0.4  \n",
       "3       20       18        0.4  \n",
       "4       21       20        0.4  \n",
       "5       20       17        0.4  \n",
       "\n",
       "[6 rows x 32 columns]"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PCR_Data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
