{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Adjust Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#import libraries\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#import data\n",
    "combinedF = os.path.join(\"C:/Users/jpcor/GoogleDrive/Data Science/Clauset's/PCR Analysis/UncleanedData/\", \"Combined_2017-2019-20_RawData_092620.csv\")  \n",
    "PCR_Data = pd.read_csv(combinedF)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "before dropping duplicates 327\n",
      "after dropping duplicates 297\n"
     ]
    }
   ],
   "source": [
    "#Drop duplicate data 10/26/20\n",
    "print('before dropping duplicates',len(PCR_Data))\n",
    "PCR_Data.drop_duplicates(keep='first',inplace=True)\n",
    "print('after dropping duplicates',len(PCR_Data))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Drop unnecessary columns\n",
    "remove=[\"timestamp\",\"user\",\"email\",\"species\",\"polymerase_catalog_number\",\"pcrid\",\n",
    "        \"template_type\",\"polymerase_brand\",\"dna_amt_units\",\"band_size_units\",'customized_salts']\n",
    "PCR_Data.drop(remove,axis=1,inplace=True)\n",
    "PCR_Data.dropna(inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Set Outcomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def outcome_bool(a):\n",
    "    '''\n",
    "    Description:\n",
    "    ------------\n",
    "    Converts Truth values for inputs right_band, wrong_band, no_bands into 1 or 0 for success or fail respectively.\n",
    "    Classifies \"dirty\" PCRs as success by only considering the \"right_band\" input.\n",
    "    \n",
    "    Parameters:\n",
    "    ------------\n",
    "    right_band; bool: True or False\n",
    "    \n",
    "    Return:\n",
    "    ------------\n",
    "    int: 1 or 0\n",
    "    '''\n",
    "    if a:\n",
    "        return 1\n",
    "    else:\n",
    "        return 0\n",
    "def outcome_bool_clean(a,b,c):\n",
    "    '''\n",
    "    Description:\n",
    "    ------------\n",
    "    Converts Truth values for inputs right_band, wrong_band, no_bands into 1 or 0 for success or fail.\n",
    "    Classifies only \"clean\" PCRs as successes by considering both right_band and wrong_band inputs.\n",
    "    \n",
    "    Parameters:\n",
    "    ------------\n",
    "    a,b,c ; bool: True or False #right_band, wrong_bands, no_bands\n",
    "    \n",
    "    Return:\n",
    "    ------------\n",
    "    int: 1 or 0\n",
    "    '''\n",
    "    if c: #no bands\n",
    "        return 0\n",
    "    elif (a and not b): #right bands only\n",
    "        return 1\n",
    "    else: #wrong bands and right bands\n",
    "        return 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "#outcome\n",
    "PCR_Data[\"outcome\"] = PCR_Data[\"right_band\"].apply(lambda x:outcome_bool(x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>primer1</th>\n",
       "      <th>primer2</th>\n",
       "      <th>amtprimer1</th>\n",
       "      <th>amtprimer2</th>\n",
       "      <th>polymerase_name</th>\n",
       "      <th>t1_min</th>\n",
       "      <th>temp1</th>\n",
       "      <th>cycles1</th>\n",
       "      <th>melting_t2_min</th>\n",
       "      <th>melting_temp2</th>\n",
       "      <th>...</th>\n",
       "      <th>t1_sec</th>\n",
       "      <th>melting_t2_sec</th>\n",
       "      <th>annealing_t2_sec</th>\n",
       "      <th>extension_t2_min.1</th>\n",
       "      <th>t5_sec</th>\n",
       "      <th>wrong_bands</th>\n",
       "      <th>right_band</th>\n",
       "      <th>no_bands</th>\n",
       "      <th>outcome</th>\n",
       "      <th>outcome_clean</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>TGCCTCCCTCAGCAAGGC</td>\n",
       "      <td>TGCCTCCCTCAGCAAGGC</td>\n",
       "      <td>0.25</td>\n",
       "      <td>0.25</td>\n",
       "      <td>Platinum PCR SuperMix High Fidelity</td>\n",
       "      <td>0</td>\n",
       "      <td>94</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>94</td>\n",
       "      <td>...</td>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>atggcgcagtgggagatgctg</td>\n",
       "      <td>CGTCGTGGTTGCCTACC</td>\n",
       "      <td>0.40</td>\n",
       "      <td>0.40</td>\n",
       "      <td>LongAmp Taq DNA Polymerase</td>\n",
       "      <td>0</td>\n",
       "      <td>94</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>94</td>\n",
       "      <td>...</td>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "      <td>45</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>TGCCTCCCTCAGCAAGGC</td>\n",
       "      <td>TGCCTCCCTCAGCAAGGC</td>\n",
       "      <td>0.25</td>\n",
       "      <td>0.25</td>\n",
       "      <td>Platinum PCR SuperMix High Fidelity</td>\n",
       "      <td>0</td>\n",
       "      <td>94</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>94</td>\n",
       "      <td>...</td>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  primer1             primer2  amtprimer1  amtprimer2  \\\n",
       "41     TGCCTCCCTCAGCAAGGC  TGCCTCCCTCAGCAAGGC        0.25        0.25   \n",
       "51  atggcgcagtgggagatgctg   CGTCGTGGTTGCCTACC        0.40        0.40   \n",
       "60     TGCCTCCCTCAGCAAGGC  TGCCTCCCTCAGCAAGGC        0.25        0.25   \n",
       "\n",
       "                        polymerase_name  t1_min  temp1  cycles1  \\\n",
       "41  Platinum PCR SuperMix High Fidelity       0     94        1   \n",
       "51           LongAmp Taq DNA Polymerase       0     94        1   \n",
       "60  Platinum PCR SuperMix High Fidelity       0     94        1   \n",
       "\n",
       "    melting_t2_min  melting_temp2  ...  t1_sec  melting_t2_sec  \\\n",
       "41               0             94  ...      30              30   \n",
       "51               0             94  ...      30              30   \n",
       "60               0             94  ...      30              30   \n",
       "\n",
       "    annealing_t2_sec  extension_t2_min.1  t5_sec  wrong_bands  right_band  \\\n",
       "41                30                  30       0         True        True   \n",
       "51                45                   0       0         True        True   \n",
       "60                30                  30       0         True        True   \n",
       "\n",
       "    no_bands  outcome  outcome_clean  \n",
       "41     False        1              0  \n",
       "51     False        1              0  \n",
       "60     False        1              0  \n",
       "\n",
       "[3 rows x 30 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Only 100%successful no nonspecific products\n",
    "PCR_Data['outcome_clean'] = PCR_Data.apply(lambda x: outcome_bool_clean(x.right_band,x.wrong_bands,x.no_bands),axis=1)\n",
    "PCR_Data[(PCR_Data['right_band']==True)&(PCR_Data['wrong_bands']==True)].head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Drop original output columns\n",
    "remove=[\"wrong_bands\",\"right_band\",\"no_bands\"]\n",
    "PCR_Data.drop(remove,axis=1,inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Data Cleaning"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "#rename incorrect column labels to correct column labels\n",
    "PCR_Data.rename(columns={'extension_t2_min.1':'extension_t2_sec','t5_sec':'t3_sec'},inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Convert time inputs to seconds\n",
    "PCR_Data[\"t1(sec)\"] = (PCR_Data[\"t1_min\"]*60) + PCR_Data[\"t1_sec\"]\n",
    "PCR_Data[\"melting_t2(sec)\"] = (PCR_Data[\"melting_t2_min\"]*60) + PCR_Data[\"melting_t2_sec\"]\n",
    "PCR_Data[\"annealing_t3(sec)\"] = (PCR_Data[\"annealing_t2_min\"]*60) + PCR_Data[\"annealing_t2_sec\"]\n",
    "PCR_Data[\"extension_t4(sec)\"] = (PCR_Data[\"extension_t2_min\"]*60) + PCR_Data[\"extension_t2_sec\"]\n",
    "PCR_Data[\"t5(sec)\"] = (PCR_Data[\"t3_min\"]*60) + PCR_Data[\"t3_sec\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "#drop old time columns\n",
    "remove=[\"t1_min\",\"t1_sec\",\"melting_t2_min\",\"melting_t2_sec\",\"annealing_t2_min\",\n",
    "        \"annealing_t2_sec\",\"extension_t2_min\",\"extension_t2_sec\",\"t3_min\",\"t3_sec\"]\n",
    "PCR_Data.drop(remove,axis=1,inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Feature Engineering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.SeqUtils import MeltingTemp as mt\n",
    "from Bio.SeqUtils import GC\n",
    "import primer3 as pr\n",
    "import melting as melt\n",
    "from NickFunctions import*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "def Tm(primer,polymerase,primerconc, option,primer2='NA'):\n",
    "    '''\n",
    "    \n",
    "    \n",
    "    Description:\n",
    "    ------------\n",
    "    Calculates primer melting temperature (Tm) based on a variety of options. Uilizes Buffer function to lookup salt concentrations for each function. Multiple options for testing.\n",
    "     NOTE: primer2 only necessary when calculating heterodimer\n",
    "    Options (default 3): \n",
    "    1 - Primer3 melting temperature\n",
    "    2 - Tm_NN BioSeqUtils\n",
    "    3 - melting (Virtually the same as IDT)\n",
    "    4 - Hairpin\n",
    "    5 - Homodimer\n",
    "    6 - Heterodimer need to input primer2\n",
    "    \n",
    "    Differences:\n",
    "\n",
    "    IDT unsure what NN method is used for calculation. Should use Owczarzy (2008) salt correction.\n",
    "\n",
    "    Primer3 uses santaLucia (1997) correction. Uses Owczarzy (2008) salt correction\n",
    "\n",
    "    BioSeqUtils uses santaLucia (2004) review correction. Uses Owczarzy (2008) salt correction\n",
    "    \n",
    "    Parameters:\n",
    "    ------------\n",
    "    primer, str: primer to calculate melting temperature for\n",
    "    polymerase, str: simplified polyerase name from \"ChangeName\" function. Used for lookup table of salt concentrations.\n",
    "    primerconc, float: from user input\n",
    "    option, int: melting temp prediction option\n",
    "    primer2, str: only needed for heterodimer calculation option.\n",
    "    \n",
    "    Return:\n",
    "    ------------\n",
    "    calculated melting temperature as float.\n",
    "    '''\n",
    "    \n",
    "    b=Buffer(polymerase)#salts in each polymerase buffer\n",
    "    Tms=[]\n",
    "    if (option==1):\n",
    "        return pr.calcTm(seq=primer,\n",
    "                                    mv_conc=b['MV_tot'],\n",
    "                                    dv_conc=b['DV_tot'],\n",
    "                                    dntp_conc=b['dNTPs'],\n",
    "                                    dna_conc=primerconc,\n",
    "                                    max_nn_length=60,\n",
    "                                    tm_method='santalucia',\n",
    "                                    salt_corrections_method=2)\n",
    "    elif (option==2):\n",
    "        return mt.Tm_NN(seq=primer,\n",
    "                                    nn_table=mt.DNA_NN4,\n",
    "                                    dnac1=primerconc,\n",
    "                                    dnac2=0,\n",
    "                                    Na=b['Na'], \n",
    "                                    K=b['K'],\n",
    "                                    Tris=b['Tris'],\n",
    "                                    Mg=b['Mg'],\n",
    "                                    dNTPs=b['dNTPs'],\n",
    "                                    saltcorr=7)\n",
    "    elif (option==3):\n",
    "        return melt.temp(primer,\n",
    "                                    DNA_c=primerconc,\n",
    "                                    Na_c=b['MV_tot'],\n",
    "                                    Mg_c=b['DV_tot'],\n",
    "                                    dNTPs_c=b['dNTPs'])\n",
    "    elif (option==4):\n",
    "        return pr.calcHairpin(seq=primer,\n",
    "                                    mv_conc=b['MV_tot'],\n",
    "                                    dv_conc=b['DV_tot'],\n",
    "                                    dntp_conc=b['dNTPs'],\n",
    "                                    dna_conc=primerconc,\n",
    "                                    temp_c=37,\n",
    "                                    max_loop=30,\n",
    "                                    output_structure=False).tm\n",
    "    elif (option==5):\n",
    "        return pr.calcHomodimer(seq=primer,\n",
    "                                    mv_conc=b['MV_tot'],\n",
    "                                    dv_conc=b['DV_tot'],\n",
    "                                    dntp_conc=b['dNTPs'],\n",
    "                                    dna_conc=primerconc,\n",
    "                                    temp_c=37,\n",
    "                                    max_loop=30,\n",
    "                                    output_structure=False).tm\n",
    "    elif (option==6):\n",
    "        return pr.calcHeterodimer(seq1=primer, seq2=primer2,\n",
    "                                    mv_conc=b['MV_tot'],\n",
    "                                    dv_conc=b['DV_tot'],\n",
    "                                    dntp_conc=b['dNTPs'],\n",
    "                                    dna_conc=primerconc,\n",
    "                                    temp_c=37,\n",
    "                                    max_loop=30,\n",
    "                                    output_structure=False).tm\n",
    "    else:\n",
    "        print(\"Not a valid option. Please input 1,2,3,4,5, or 6\")\n",
    "        return 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>primer1</th>\n",
       "      <th>primer2</th>\n",
       "      <th>amtprimer1</th>\n",
       "      <th>amtprimer2</th>\n",
       "      <th>polymerase_name</th>\n",
       "      <th>temp1</th>\n",
       "      <th>cycles1</th>\n",
       "      <th>melting_temp2</th>\n",
       "      <th>annealing_temp2</th>\n",
       "      <th>extension_temp2</th>\n",
       "      <th>...</th>\n",
       "      <th>cycles3</th>\n",
       "      <th>band_size</th>\n",
       "      <th>dna_amt</th>\n",
       "      <th>outcome</th>\n",
       "      <th>outcome_clean</th>\n",
       "      <th>t1(sec)</th>\n",
       "      <th>melting_t2(sec)</th>\n",
       "      <th>annealing_t3(sec)</th>\n",
       "      <th>extension_t4(sec)</th>\n",
       "      <th>t5(sec)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCTTCACATTCTCAGGAACTCTCC</td>\n",
       "      <td>GCAGGGCGAGGGGCAGG</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.2</td>\n",
       "      <td>SuperMix</td>\n",
       "      <td>94</td>\n",
       "      <td>1</td>\n",
       "      <td>94</td>\n",
       "      <td>60</td>\n",
       "      <td>72</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>369.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "      <td>600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCTTCACATTCTCAGGAACTCTCC</td>\n",
       "      <td>GGCATGGTTCTCTTCTCTGC</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.2</td>\n",
       "      <td>SuperMix</td>\n",
       "      <td>94</td>\n",
       "      <td>1</td>\n",
       "      <td>94</td>\n",
       "      <td>60</td>\n",
       "      <td>72</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>369.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "      <td>600</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    primer1               primer2  amtprimer1  amtprimer2  \\\n",
       "0  GCTTCACATTCTCAGGAACTCTCC     GCAGGGCGAGGGGCAGG         0.2         0.2   \n",
       "1  GCTTCACATTCTCAGGAACTCTCC  GGCATGGTTCTCTTCTCTGC         0.2         0.2   \n",
       "\n",
       "  polymerase_name  temp1  cycles1  melting_temp2  annealing_temp2  \\\n",
       "0        SuperMix     94        1             94               60   \n",
       "1        SuperMix     94        1             94               60   \n",
       "\n",
       "   extension_temp2  ...  cycles3  band_size  dna_amt  outcome  outcome_clean  \\\n",
       "0               72  ...        1      369.0      0.0        0              0   \n",
       "1               72  ...        1      369.0      0.0        0              0   \n",
       "\n",
       "   t1(sec)  melting_t2(sec)  annealing_t3(sec)  extension_t4(sec)  t5(sec)  \n",
       "0       30               30                 30                 30      600  \n",
       "1       30               30                 30                 30      600  \n",
       "\n",
       "[2 rows x 22 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Change polymerase name to generalized form for Buffer dictionary\n",
    "PCR_Data[\"polymerase_name\"]=PCR_Data['polymerase_name'].apply(lambda x: ChangeName(x))\n",
    "PCR_Data.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Strip blank spaces from primers\n",
    "PCR_Data[\"primer1\"] = PCR_Data[\"primer1\"].apply(lambda x: x.replace(' ',''))# remove all white space including middle of string\n",
    "PCR_Data[\"primer2\"] = PCR_Data[\"primer2\"].apply(lambda x: x.replace(' ',''))# remove all white space including middle of string\n",
    "#Ensure all primers are upper case\n",
    "PCR_Data[\"primer1\"] = PCR_Data[\"primer1\"].apply(lambda x: x.upper())\n",
    "PCR_Data[\"primer2\"] = PCR_Data[\"primer2\"].apply(lambda x: x.upper())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Add Primer GC content\n",
    "PCR_Data[\"primer1_GC\"] = PCR_Data[\"primer1\"].apply(lambda x: GC(x))\n",
    "PCR_Data[\"primer2_GC\"] = PCR_Data[\"primer2\"].apply(lambda x: GC(x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Add Primer Lengths\n",
    "PCR_Data[\"P1_len\"] = PCR_Data[\"primer1\"].apply(lambda x: len(x))\n",
    "PCR_Data[\"P2_len\"] = PCR_Data[\"primer2\"].apply(lambda x: len(x))\n",
    "#Remove primers with length greater than 60 bioinformatics algorithms do not work well when greater than 60\n",
    "longPs = PCR_Data[(PCR_Data[\"P1_len\"]>60) | (PCR_Data[\"P2_len\"]>60)].index \n",
    "PCR_Data.drop(longPs,inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Add Primer melting temps\n",
    "#Primer 3 no correction original model\n",
    "PCR_Data[\"Tm_p1_p3orig\"]=PCR_Data.apply(lambda x: pr.calcTm(x.primer1),axis=1)\n",
    "PCR_Data[\"Tm_p2_p3orig\"]=PCR_Data.apply(lambda x: pr.calcTm(x.primer2),axis=1)\n",
    "#Primer3\n",
    "PCR_Data[\"Tm_p1_p3\"]=PCR_Data.apply(lambda x: Tm(primer=x.primer1,polymerase=x.polymerase_name,primerconc=(x.amtprimer1+x.amtprimer2),option=1),axis=1)\n",
    "PCR_Data[\"Tm_p2_p3\"]=PCR_Data.apply(lambda x: Tm(primer=x.primer2,polymerase=x.polymerase_name,primerconc=(x.amtprimer1+x.amtprimer2),option=1),axis=1)\n",
    "#BioseqUtils\n",
    "PCR_Data[\"Tm_p1_BSU\"]=PCR_Data.apply(lambda x: Tm(primer=x.primer1,polymerase=x.polymerase_name,primerconc=(x.amtprimer1+x.amtprimer2),option=2),axis=1)\n",
    "PCR_Data[\"Tm_p2_BSU\"]=PCR_Data.apply(lambda x: Tm(primer=x.primer2,polymerase=x.polymerase_name,primerconc=(x.amtprimer1+x.amtprimer2),option=2),axis=1)\n",
    "#Melting (IDT analog)\n",
    "PCR_Data[\"Tm_p1_IDT\"]=PCR_Data.apply(lambda x: Tm(primer=x.primer1,polymerase=x.polymerase_name,primerconc=(x.amtprimer1+x.amtprimer2),option=3),axis=1)\n",
    "PCR_Data[\"Tm_p2_IDT\"]=PCR_Data.apply(lambda x: Tm(primer=x.primer2,polymerase=x.polymerase_name,primerconc=(x.amtprimer1+x.amtprimer2),option=3),axis=1)\n",
    "#Average\n",
    "#PCR_Data[\"Tm_p1_Avg\"]=(PCR_Data[\"Tm_p1_p3\"]+PCR_Data[\"Tm_p1_BSU\"]+PCR_Data[\"Tm_p1_IDT\"])/3\n",
    "#PCR_Data[\"Tm_p2_Avg\"]=(PCR_Data[\"Tm_p2_p3\"]+PCR_Data[\"Tm_p2_BSU\"]+PCR_Data[\"Tm_p2_IDT\"])/3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "#difference in primer melting temps\n",
    "PCR_Data[\"Tm_dif_IDT\"]=abs(PCR_Data[\"Tm_p1_IDT\"]-PCR_Data[\"Tm_p2_IDT\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Add if hairpin structure is predicted\n",
    "PCR_Data[\"HP_P1\"] = PCR_Data.apply(lambda x: Tm(primer=x.primer1,polymerase=x.polymerase_name,primerconc=(x.amtprimer1+x.amtprimer2),option=4),axis=1)\n",
    "PCR_Data[\"HP_P2\"] = PCR_Data.apply(lambda x: Tm(primer=x.primer2,polymerase=x.polymerase_name,primerconc=(x.amtprimer1+x.amtprimer2),option=4),axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Combine hairpins to one feature that has the highest hairpin temp\n",
    "#if first statement is true return 2nd value first statement is false return third value.\n",
    "PCR_Data[\"HPin_Tm_Max\"] = np.where(PCR_Data[\"HP_P1\"]>PCR_Data[\"HP_P2\"],PCR_Data[\"HP_P1\"],PCR_Data[\"HP_P2\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "290.0"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(PCR_Data.size)/len(PCR_Data.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Calculate Heterodimer\n",
    "PCR_Data[\"Heterodimer_tm\"] = PCR_Data.apply(lambda x : Tm(primer=x.primer1,primer2=x.primer2,polymerase=x.polymerase_name,primerconc=(x.amtprimer1+x.amtprimer2),option=6),axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Calculate Homodimer\n",
    "PCR_Data[\"Homodimer_tm1\"] = PCR_Data.apply(lambda x : Tm(primer=x.primer1,polymerase=x.polymerase_name,primerconc=(x.amtprimer1+x.amtprimer2),option=5),axis=1)\n",
    "PCR_Data[\"Homodimer_tm2\"] = PCR_Data.apply(lambda x : Tm(primer=x.primer2,polymerase=x.polymerase_name,primerconc=(x.amtprimer1+x.amtprimer2),option=5),axis=1)\n",
    "PCR_Data[\"Homodimer_maxTm\"] = np.where(PCR_Data[\"Homodimer_tm1\"]>PCR_Data[\"Homodimer_tm2\"],PCR_Data[\"Homodimer_tm1\"],PCR_Data[\"Homodimer_tm2\"])\n",
    "PCR_Data[\"Homodimer_minTm\"] = np.where(PCR_Data[\"Homodimer_tm1\"]<PCR_Data[\"Homodimer_tm2\"],PCR_Data[\"Homodimer_tm1\"],PCR_Data[\"Homodimer_tm2\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "def changeTm(Tm):\n",
    "    '''\n",
    "    Tms of less than or equal to zero mean there was no significant structure. To help the RF make these temps all 0\n",
    "    '''\n",
    "    if Tm <0:\n",
    "        return 0\n",
    "    else:\n",
    "        return Tm\n",
    "PCR_Data[\"Homodimer_maxTm\"]=PCR_Data[\"Homodimer_maxTm\"].apply(lambda x: changeTm(x))\n",
    "PCR_Data[\"Heterodimer_tm\"]=PCR_Data[\"Heterodimer_tm\"].apply(lambda x: changeTm(x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "PCR_Data['MaxTm_IDT']=np.where(PCR_Data[\"Tm_p1_IDT\"]>PCR_Data[\"Tm_p2_IDT\"],PCR_Data[\"Tm_p1_IDT\"],PCR_Data[\"Tm_p2_IDT\"])\n",
    "PCR_Data['MinTm_IDT']=np.where(PCR_Data[\"Tm_p1_IDT\"]<PCR_Data[\"Tm_p2_IDT\"],PCR_Data[\"Tm_p1_IDT\"],PCR_Data[\"Tm_p2_IDT\"])\n",
    "PCR_Data['MaxGC']=np.where(PCR_Data[\"primer1_GC\"]>PCR_Data[\"primer2_GC\"],PCR_Data[\"primer1_GC\"],PCR_Data[\"primer2_GC\"])\n",
    "PCR_Data['MinGC']=np.where(PCR_Data[\"primer1_GC\"]<PCR_Data[\"primer2_GC\"],PCR_Data[\"primer1_GC\"],PCR_Data[\"primer2_GC\"])\n",
    "PCR_Data['MaxPlen']=np.where(PCR_Data[\"P1_len\"]>PCR_Data[\"P2_len\"],PCR_Data[\"P1_len\"],PCR_Data[\"P2_len\"])\n",
    "PCR_Data['MinPlen']=np.where(PCR_Data[\"P1_len\"]<PCR_Data[\"P2_len\"],PCR_Data[\"P1_len\"],PCR_Data[\"P2_len\"])\n",
    "PCR_Data['TotPrimer']=PCR_Data['amtprimer1']+PCR_Data['amtprimer2']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**GC Clamp:**\n",
    "\n",
    "According to Top Tip Bio (TTP) (https://toptipbio.com/gc-clamp-pcr/) GC Clamp is defined by a G or C in the last 5 nucleotides on the 3' end of the primer.\n",
    "\n",
    "Strength table for NN model taken from Khandelwahl (2010)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "class GC_Clamp_features(object):\n",
    "    '''\n",
    "    Description:\n",
    "    ------------\n",
    "    Class obtains features from GC clamp such as strength and num_GC \n",
    "    NN model strength table taken from \n",
    "    Khandelwal G, Bhyravabhotla J. A Phenomenological Model for Predicting Melting Temperatures of DNA Sequences. PLOS ONE. 2010;5: e12433. doi:10.1371/journal.pone.0012433\n",
    "    '''\n",
    "    #initialize class variables\n",
    "    clamp = \"\"\n",
    "    num_GC = 0\n",
    "    Strength={'GC': 13,'CC': 11,'GG': 11,'CG': 10,'AC': 10,'TC': 8,'AG': 8,'TG': 7,'GT': 10,'CT': 8,'GA': 8,'CA': 7,'AT': 7,'TT': 5,'AA': 5,'TA': 4}\n",
    "    Score=0\n",
    "    #GC_Clamp class constructor\n",
    "    def __init__(self,primer):\n",
    "        self.clamp = primer[-5:]\n",
    "        for nt in range(len(self.clamp)-1):#inrement by 1 nucleotide at a time\n",
    "            NN=self.clamp[nt]+self.clamp[nt+1]#grab dinucleotides\n",
    "            self.Score+=self.Strength[NN]#GC clamp strength score\n",
    "        for nt in range(len(self.clamp)):\n",
    "            if self.clamp[nt]==\"G\" or self.clamp[nt]==\"C\":\n",
    "                self.num_GC+=1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TAGCA\n",
      "2\n",
      "32\n"
     ]
    }
   ],
   "source": [
    "clamp = GC_Clamp_features(\"ATACGTACGATAGCA\")\n",
    "print(clamp.clamp)\n",
    "print(clamp.num_GC)\n",
    "print(clamp.Score)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Create GC_Clamp_features Score\n",
    "PCR_Data[\"GC_Clamp1\"] = PCR_Data[\"primer1\"].apply(lambda x: GC_Clamp_features(x).Score)\n",
    "PCR_Data[\"GC_Clamp2\"] = PCR_Data[\"primer2\"].apply(lambda x: GC_Clamp_features(x).Score)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finalize GC_Clamp features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Take the lowest GC in clamp and positional weight for p1 and p2\n",
    "PCR_Data[\"GC_ClampMin\"] = np.where(PCR_Data[\"GC_Clamp1\"]<PCR_Data[\"GC_Clamp2\"],PCR_Data[\"GC_Clamp1\"],PCR_Data[\"GC_Clamp2\"])\n",
    "PCR_Data[\"GC_ClampMax\"] = np.where(PCR_Data[\"GC_Clamp1\"]>PCR_Data[\"GC_Clamp2\"],PCR_Data[\"GC_Clamp1\"],PCR_Data[\"GC_Clamp2\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>temp1</th>\n",
       "      <th>melting_temp2</th>\n",
       "      <th>annealing_temp2</th>\n",
       "      <th>extension_temp2</th>\n",
       "      <th>cycles2</th>\n",
       "      <th>temp3</th>\n",
       "      <th>band_size</th>\n",
       "      <th>dna_amt</th>\n",
       "      <th>outcome</th>\n",
       "      <th>outcome_clean</th>\n",
       "      <th>...</th>\n",
       "      <th>Homodimer_minTm</th>\n",
       "      <th>MaxTm_IDT</th>\n",
       "      <th>MinTm_IDT</th>\n",
       "      <th>MaxGC</th>\n",
       "      <th>MinGC</th>\n",
       "      <th>MaxPlen</th>\n",
       "      <th>MinPlen</th>\n",
       "      <th>TotPrimer</th>\n",
       "      <th>GC_ClampMin</th>\n",
       "      <th>GC_ClampMax</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>94</td>\n",
       "      <td>94</td>\n",
       "      <td>60</td>\n",
       "      <td>72</td>\n",
       "      <td>44</td>\n",
       "      <td>72</td>\n",
       "      <td>369.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>-68.100370</td>\n",
       "      <td>58.394749</td>\n",
       "      <td>55.977067</td>\n",
       "      <td>82.352941</td>\n",
       "      <td>50.000000</td>\n",
       "      <td>24</td>\n",
       "      <td>17</td>\n",
       "      <td>0.4</td>\n",
       "      <td>35</td>\n",
       "      <td>39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>94</td>\n",
       "      <td>94</td>\n",
       "      <td>60</td>\n",
       "      <td>72</td>\n",
       "      <td>44</td>\n",
       "      <td>72</td>\n",
       "      <td>369.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>-68.100370</td>\n",
       "      <td>55.977067</td>\n",
       "      <td>52.726430</td>\n",
       "      <td>55.000000</td>\n",
       "      <td>50.000000</td>\n",
       "      <td>24</td>\n",
       "      <td>20</td>\n",
       "      <td>0.4</td>\n",
       "      <td>35</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>94</td>\n",
       "      <td>94</td>\n",
       "      <td>60</td>\n",
       "      <td>72</td>\n",
       "      <td>44</td>\n",
       "      <td>72</td>\n",
       "      <td>369.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>-68.100370</td>\n",
       "      <td>59.131380</td>\n",
       "      <td>55.977067</td>\n",
       "      <td>70.000000</td>\n",
       "      <td>50.000000</td>\n",
       "      <td>24</td>\n",
       "      <td>20</td>\n",
       "      <td>0.4</td>\n",
       "      <td>35</td>\n",
       "      <td>39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>94</td>\n",
       "      <td>94</td>\n",
       "      <td>60</td>\n",
       "      <td>72</td>\n",
       "      <td>44</td>\n",
       "      <td>72</td>\n",
       "      <td>369.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>-57.711322</td>\n",
       "      <td>58.394749</td>\n",
       "      <td>55.507339</td>\n",
       "      <td>82.352941</td>\n",
       "      <td>57.142857</td>\n",
       "      <td>21</td>\n",
       "      <td>17</td>\n",
       "      <td>0.4</td>\n",
       "      <td>39</td>\n",
       "      <td>39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>94</td>\n",
       "      <td>94</td>\n",
       "      <td>60</td>\n",
       "      <td>72</td>\n",
       "      <td>44</td>\n",
       "      <td>72</td>\n",
       "      <td>369.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>-25.549291</td>\n",
       "      <td>55.507339</td>\n",
       "      <td>52.726430</td>\n",
       "      <td>57.142857</td>\n",
       "      <td>55.000000</td>\n",
       "      <td>21</td>\n",
       "      <td>20</td>\n",
       "      <td>0.4</td>\n",
       "      <td>36</td>\n",
       "      <td>39</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 35 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   temp1  melting_temp2  annealing_temp2  extension_temp2  cycles2  temp3  \\\n",
       "0     94             94               60               72       44     72   \n",
       "1     94             94               60               72       44     72   \n",
       "2     94             94               60               72       44     72   \n",
       "3     94             94               60               72       44     72   \n",
       "4     94             94               60               72       44     72   \n",
       "\n",
       "   band_size  dna_amt  outcome  outcome_clean  ...  Homodimer_minTm  \\\n",
       "0      369.0      0.0        0              0  ...       -68.100370   \n",
       "1      369.0      0.0        0              0  ...       -68.100370   \n",
       "2      369.0      0.0        0              0  ...       -68.100370   \n",
       "3      369.0      0.0        1              1  ...       -57.711322   \n",
       "4      369.0      0.0        1              1  ...       -25.549291   \n",
       "\n",
       "   MaxTm_IDT  MinTm_IDT      MaxGC      MinGC  MaxPlen  MinPlen  TotPrimer  \\\n",
       "0  58.394749  55.977067  82.352941  50.000000       24       17        0.4   \n",
       "1  55.977067  52.726430  55.000000  50.000000       24       20        0.4   \n",
       "2  59.131380  55.977067  70.000000  50.000000       24       20        0.4   \n",
       "3  58.394749  55.507339  82.352941  57.142857       21       17        0.4   \n",
       "4  55.507339  52.726430  57.142857  55.000000       21       20        0.4   \n",
       "\n",
       "   GC_ClampMin  GC_ClampMax  \n",
       "0           35           39  \n",
       "1           35           36  \n",
       "2           35           39  \n",
       "3           39           39  \n",
       "4           36           39  \n",
       "\n",
       "[5 rows x 35 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Drop GC_CLamp columns used for analysis and primers because we are done pulling data from them\n",
    "remove=[\"GC_Clamp1\",\"GC_Clamp2\",\"primer1\",\"primer2\",\"polymerase_name\",\"cycles1\",\"cycles3\",\"HP_P1\",\"HP_P2\",\"Homodimer_tm1\",\"Homodimer_tm2\",\"Tm_p1_IDT\",\"Tm_p2_IDT\",\"primer1_GC\",\"primer2_GC\",\"P1_len\",\"P2_len\",'amtprimer1','amtprimer2']\n",
    "PCR_Data.drop(remove,axis=1,inplace=True)\n",
    "PCR_Data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.0"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PCR_Data['dna_amt'].min()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>outcome</th>\n",
       "      <th>outcome_clean</th>\n",
       "      <th>temp1</th>\n",
       "      <th>melting_temp2</th>\n",
       "      <th>annealing_temp2</th>\n",
       "      <th>extension_temp2</th>\n",
       "      <th>cycles2</th>\n",
       "      <th>temp3</th>\n",
       "      <th>band_size</th>\n",
       "      <th>dna_amt</th>\n",
       "      <th>...</th>\n",
       "      <th>Homodimer_maxTm</th>\n",
       "      <th>GC_ClampMin</th>\n",
       "      <th>GC_ClampMax</th>\n",
       "      <th>MaxTm_IDT</th>\n",
       "      <th>MinTm_IDT</th>\n",
       "      <th>MaxGC</th>\n",
       "      <th>MinGC</th>\n",
       "      <th>MaxPlen</th>\n",
       "      <th>MinPlen</th>\n",
       "      <th>TotPrimer</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>94</td>\n",
       "      <td>94</td>\n",
       "      <td>54</td>\n",
       "      <td>72</td>\n",
       "      <td>30</td>\n",
       "      <td>72</td>\n",
       "      <td>400.0</td>\n",
       "      <td>200.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>37</td>\n",
       "      <td>37</td>\n",
       "      <td>55.852404</td>\n",
       "      <td>55.852404</td>\n",
       "      <td>66.666667</td>\n",
       "      <td>66.666667</td>\n",
       "      <td>18</td>\n",
       "      <td>18</td>\n",
       "      <td>0.50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>94</td>\n",
       "      <td>94</td>\n",
       "      <td>56</td>\n",
       "      <td>65</td>\n",
       "      <td>30</td>\n",
       "      <td>65</td>\n",
       "      <td>2.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>33</td>\n",
       "      <td>35</td>\n",
       "      <td>59.846603</td>\n",
       "      <td>51.410944</td>\n",
       "      <td>64.705882</td>\n",
       "      <td>61.904762</td>\n",
       "      <td>21</td>\n",
       "      <td>17</td>\n",
       "      <td>0.80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>94</td>\n",
       "      <td>94</td>\n",
       "      <td>54</td>\n",
       "      <td>72</td>\n",
       "      <td>30</td>\n",
       "      <td>72</td>\n",
       "      <td>400.0</td>\n",
       "      <td>200.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>37</td>\n",
       "      <td>37</td>\n",
       "      <td>55.852404</td>\n",
       "      <td>55.852404</td>\n",
       "      <td>66.666667</td>\n",
       "      <td>66.666667</td>\n",
       "      <td>18</td>\n",
       "      <td>18</td>\n",
       "      <td>0.50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>98</td>\n",
       "      <td>98</td>\n",
       "      <td>71</td>\n",
       "      <td>72</td>\n",
       "      <td>30</td>\n",
       "      <td>72</td>\n",
       "      <td>1100.0</td>\n",
       "      <td>200.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>35</td>\n",
       "      <td>41</td>\n",
       "      <td>57.363314</td>\n",
       "      <td>57.143668</td>\n",
       "      <td>54.545455</td>\n",
       "      <td>52.173913</td>\n",
       "      <td>23</td>\n",
       "      <td>22</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>98</td>\n",
       "      <td>98</td>\n",
       "      <td>69</td>\n",
       "      <td>72</td>\n",
       "      <td>30</td>\n",
       "      <td>72</td>\n",
       "      <td>1100.0</td>\n",
       "      <td>200.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>35</td>\n",
       "      <td>41</td>\n",
       "      <td>57.363314</td>\n",
       "      <td>57.143668</td>\n",
       "      <td>54.545455</td>\n",
       "      <td>52.173913</td>\n",
       "      <td>23</td>\n",
       "      <td>22</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>94</td>\n",
       "      <td>94</td>\n",
       "      <td>60</td>\n",
       "      <td>72</td>\n",
       "      <td>35</td>\n",
       "      <td>72</td>\n",
       "      <td>1060.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>32</td>\n",
       "      <td>39</td>\n",
       "      <td>59.772982</td>\n",
       "      <td>58.936270</td>\n",
       "      <td>48.000000</td>\n",
       "      <td>36.363636</td>\n",
       "      <td>33</td>\n",
       "      <td>25</td>\n",
       "      <td>0.71</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>94</td>\n",
       "      <td>94</td>\n",
       "      <td>60</td>\n",
       "      <td>72</td>\n",
       "      <td>35</td>\n",
       "      <td>72</td>\n",
       "      <td>1060.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>30</td>\n",
       "      <td>32</td>\n",
       "      <td>59.772982</td>\n",
       "      <td>59.178054</td>\n",
       "      <td>44.444444</td>\n",
       "      <td>36.363636</td>\n",
       "      <td>33</td>\n",
       "      <td>27</td>\n",
       "      <td>0.71</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>98</td>\n",
       "      <td>98</td>\n",
       "      <td>69</td>\n",
       "      <td>72</td>\n",
       "      <td>30</td>\n",
       "      <td>72</td>\n",
       "      <td>1060.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>35</td>\n",
       "      <td>41</td>\n",
       "      <td>57.890807</td>\n",
       "      <td>57.143668</td>\n",
       "      <td>54.545455</td>\n",
       "      <td>50.000000</td>\n",
       "      <td>24</td>\n",
       "      <td>22</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>98</td>\n",
       "      <td>98</td>\n",
       "      <td>65</td>\n",
       "      <td>72</td>\n",
       "      <td>35</td>\n",
       "      <td>72</td>\n",
       "      <td>4200.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>34</td>\n",
       "      <td>40</td>\n",
       "      <td>55.107337</td>\n",
       "      <td>52.831542</td>\n",
       "      <td>70.588235</td>\n",
       "      <td>63.157895</td>\n",
       "      <td>19</td>\n",
       "      <td>17</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>98</td>\n",
       "      <td>98</td>\n",
       "      <td>63</td>\n",
       "      <td>72</td>\n",
       "      <td>35</td>\n",
       "      <td>72</td>\n",
       "      <td>1668.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>13.015275</td>\n",
       "      <td>34</td>\n",
       "      <td>44</td>\n",
       "      <td>56.151147</td>\n",
       "      <td>54.609355</td>\n",
       "      <td>80.000000</td>\n",
       "      <td>44.000000</td>\n",
       "      <td>25</td>\n",
       "      <td>15</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>189</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>95</td>\n",
       "      <td>95</td>\n",
       "      <td>55</td>\n",
       "      <td>72</td>\n",
       "      <td>35</td>\n",
       "      <td>72</td>\n",
       "      <td>170.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>35</td>\n",
       "      <td>42</td>\n",
       "      <td>50.934238</td>\n",
       "      <td>50.215875</td>\n",
       "      <td>61.111111</td>\n",
       "      <td>52.631579</td>\n",
       "      <td>19</td>\n",
       "      <td>18</td>\n",
       "      <td>0.80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>95</td>\n",
       "      <td>95</td>\n",
       "      <td>55</td>\n",
       "      <td>72</td>\n",
       "      <td>35</td>\n",
       "      <td>72</td>\n",
       "      <td>180.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>36</td>\n",
       "      <td>42</td>\n",
       "      <td>55.697386</td>\n",
       "      <td>50.215875</td>\n",
       "      <td>66.666667</td>\n",
       "      <td>52.631579</td>\n",
       "      <td>19</td>\n",
       "      <td>18</td>\n",
       "      <td>0.80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>95</td>\n",
       "      <td>95</td>\n",
       "      <td>55</td>\n",
       "      <td>72</td>\n",
       "      <td>35</td>\n",
       "      <td>72</td>\n",
       "      <td>162.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>39</td>\n",
       "      <td>42</td>\n",
       "      <td>56.841631</td>\n",
       "      <td>50.215875</td>\n",
       "      <td>63.157895</td>\n",
       "      <td>52.631579</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>0.80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>192</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>95</td>\n",
       "      <td>95</td>\n",
       "      <td>55</td>\n",
       "      <td>72</td>\n",
       "      <td>35</td>\n",
       "      <td>72</td>\n",
       "      <td>194.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>33</td>\n",
       "      <td>42</td>\n",
       "      <td>52.174678</td>\n",
       "      <td>50.215875</td>\n",
       "      <td>52.631579</td>\n",
       "      <td>50.000000</td>\n",
       "      <td>20</td>\n",
       "      <td>19</td>\n",
       "      <td>0.80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>95</td>\n",
       "      <td>95</td>\n",
       "      <td>55</td>\n",
       "      <td>72</td>\n",
       "      <td>35</td>\n",
       "      <td>72</td>\n",
       "      <td>200.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>40</td>\n",
       "      <td>42</td>\n",
       "      <td>57.862494</td>\n",
       "      <td>50.215875</td>\n",
       "      <td>72.222222</td>\n",
       "      <td>52.631579</td>\n",
       "      <td>19</td>\n",
       "      <td>18</td>\n",
       "      <td>0.80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>198</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>95</td>\n",
       "      <td>95</td>\n",
       "      <td>55</td>\n",
       "      <td>72</td>\n",
       "      <td>35</td>\n",
       "      <td>72</td>\n",
       "      <td>200.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>35</td>\n",
       "      <td>42</td>\n",
       "      <td>58.483348</td>\n",
       "      <td>50.215875</td>\n",
       "      <td>54.166667</td>\n",
       "      <td>52.631579</td>\n",
       "      <td>24</td>\n",
       "      <td>19</td>\n",
       "      <td>0.80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>95</td>\n",
       "      <td>95</td>\n",
       "      <td>55</td>\n",
       "      <td>72</td>\n",
       "      <td>35</td>\n",
       "      <td>72</td>\n",
       "      <td>184.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.286358</td>\n",
       "      <td>34</td>\n",
       "      <td>42</td>\n",
       "      <td>58.121724</td>\n",
       "      <td>50.215875</td>\n",
       "      <td>56.521739</td>\n",
       "      <td>52.631579</td>\n",
       "      <td>23</td>\n",
       "      <td>19</td>\n",
       "      <td>0.80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>201</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>95</td>\n",
       "      <td>95</td>\n",
       "      <td>55</td>\n",
       "      <td>72</td>\n",
       "      <td>35</td>\n",
       "      <td>72</td>\n",
       "      <td>400.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>30</td>\n",
       "      <td>42</td>\n",
       "      <td>52.741813</td>\n",
       "      <td>50.215875</td>\n",
       "      <td>61.111111</td>\n",
       "      <td>52.631579</td>\n",
       "      <td>19</td>\n",
       "      <td>18</td>\n",
       "      <td>0.80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>218</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>95</td>\n",
       "      <td>95</td>\n",
       "      <td>69</td>\n",
       "      <td>72</td>\n",
       "      <td>34</td>\n",
       "      <td>72</td>\n",
       "      <td>2725.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>...</td>\n",
       "      <td>25.862664</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>70.716405</td>\n",
       "      <td>68.724542</td>\n",
       "      <td>47.916667</td>\n",
       "      <td>47.826087</td>\n",
       "      <td>48</td>\n",
       "      <td>46</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>305</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>95</td>\n",
       "      <td>95</td>\n",
       "      <td>55</td>\n",
       "      <td>72</td>\n",
       "      <td>35</td>\n",
       "      <td>72</td>\n",
       "      <td>184.0</td>\n",
       "      <td>1200.0</td>\n",
       "      <td>...</td>\n",
       "      <td>16.032730</td>\n",
       "      <td>34</td>\n",
       "      <td>38</td>\n",
       "      <td>72.484392</td>\n",
       "      <td>62.167654</td>\n",
       "      <td>56.521739</td>\n",
       "      <td>52.380952</td>\n",
       "      <td>42</td>\n",
       "      <td>23</td>\n",
       "      <td>20.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>316</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>95</td>\n",
       "      <td>95</td>\n",
       "      <td>60</td>\n",
       "      <td>72</td>\n",
       "      <td>35</td>\n",
       "      <td>72</td>\n",
       "      <td>370.0</td>\n",
       "      <td>1200.0</td>\n",
       "      <td>...</td>\n",
       "      <td>13.580544</td>\n",
       "      <td>33</td>\n",
       "      <td>41</td>\n",
       "      <td>74.024399</td>\n",
       "      <td>60.885243</td>\n",
       "      <td>54.761905</td>\n",
       "      <td>48.148148</td>\n",
       "      <td>42</td>\n",
       "      <td>27</td>\n",
       "      <td>20.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>323</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>95</td>\n",
       "      <td>95</td>\n",
       "      <td>60</td>\n",
       "      <td>72</td>\n",
       "      <td>36</td>\n",
       "      <td>72</td>\n",
       "      <td>502.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>...</td>\n",
       "      <td>14.014201</td>\n",
       "      <td>28</td>\n",
       "      <td>33</td>\n",
       "      <td>73.138543</td>\n",
       "      <td>61.899024</td>\n",
       "      <td>52.380952</td>\n",
       "      <td>44.827586</td>\n",
       "      <td>42</td>\n",
       "      <td>29</td>\n",
       "      <td>20.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>325</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>95</td>\n",
       "      <td>95</td>\n",
       "      <td>60</td>\n",
       "      <td>72</td>\n",
       "      <td>36</td>\n",
       "      <td>72</td>\n",
       "      <td>427.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>...</td>\n",
       "      <td>36.755084</td>\n",
       "      <td>34</td>\n",
       "      <td>41</td>\n",
       "      <td>72.642995</td>\n",
       "      <td>63.627070</td>\n",
       "      <td>51.162791</td>\n",
       "      <td>46.428571</td>\n",
       "      <td>43</td>\n",
       "      <td>28</td>\n",
       "      <td>20.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>326</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>95</td>\n",
       "      <td>95</td>\n",
       "      <td>60</td>\n",
       "      <td>72</td>\n",
       "      <td>36</td>\n",
       "      <td>72</td>\n",
       "      <td>457.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>...</td>\n",
       "      <td>32.053479</td>\n",
       "      <td>30</td>\n",
       "      <td>31</td>\n",
       "      <td>71.738337</td>\n",
       "      <td>58.776374</td>\n",
       "      <td>45.833333</td>\n",
       "      <td>44.897959</td>\n",
       "      <td>49</td>\n",
       "      <td>24</td>\n",
       "      <td>20.00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>24 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     outcome  outcome_clean  temp1  melting_temp2  annealing_temp2  \\\n",
       "41         1              0     94             94               54   \n",
       "51         1              0     94             94               56   \n",
       "60         1              0     94             94               54   \n",
       "118        1              0     98             98               71   \n",
       "119        1              0     98             98               69   \n",
       "122        1              0     94             94               60   \n",
       "123        1              0     94             94               60   \n",
       "124        1              0     98             98               69   \n",
       "135        1              0     98             98               65   \n",
       "140        1              0     98             98               63   \n",
       "189        1              0     95             95               55   \n",
       "190        1              0     95             95               55   \n",
       "191        1              0     95             95               55   \n",
       "192        1              0     95             95               55   \n",
       "193        1              0     95             95               55   \n",
       "198        1              0     95             95               55   \n",
       "200        1              0     95             95               55   \n",
       "201        1              0     95             95               55   \n",
       "218        1              0     95             95               69   \n",
       "305        1              0     95             95               55   \n",
       "316        1              0     95             95               60   \n",
       "323        1              0     95             95               60   \n",
       "325        1              0     95             95               60   \n",
       "326        1              0     95             95               60   \n",
       "\n",
       "     extension_temp2  cycles2  temp3  band_size  dna_amt  ...  \\\n",
       "41                72       30     72      400.0    200.0  ...   \n",
       "51                65       30     65        2.5      0.0  ...   \n",
       "60                72       30     72      400.0    200.0  ...   \n",
       "118               72       30     72     1100.0    200.0  ...   \n",
       "119               72       30     72     1100.0    200.0  ...   \n",
       "122               72       35     72     1060.0     20.0  ...   \n",
       "123               72       35     72     1060.0     20.0  ...   \n",
       "124               72       30     72     1060.0     20.0  ...   \n",
       "135               72       35     72     4200.0      0.0  ...   \n",
       "140               72       35     72     1668.0      0.0  ...   \n",
       "189               72       35     72      170.0   1000.0  ...   \n",
       "190               72       35     72      180.0   1000.0  ...   \n",
       "191               72       35     72      162.0   1000.0  ...   \n",
       "192               72       35     72      194.0   1000.0  ...   \n",
       "193               72       35     72      200.0   1000.0  ...   \n",
       "198               72       35     72      200.0   1000.0  ...   \n",
       "200               72       35     72      184.0   1000.0  ...   \n",
       "201               72       35     72      400.0   1000.0  ...   \n",
       "218               72       34     72     2725.0     50.0  ...   \n",
       "305               72       35     72      184.0   1200.0  ...   \n",
       "316               72       35     72      370.0   1200.0  ...   \n",
       "323               72       36     72      502.0   1000.0  ...   \n",
       "325               72       36     72      427.0   1000.0  ...   \n",
       "326               72       36     72      457.0   1000.0  ...   \n",
       "\n",
       "     Homodimer_maxTm  GC_ClampMin  GC_ClampMax  MaxTm_IDT  MinTm_IDT  \\\n",
       "41          0.000000           37           37  55.852404  55.852404   \n",
       "51          0.000000           33           35  59.846603  51.410944   \n",
       "60          0.000000           37           37  55.852404  55.852404   \n",
       "118         0.000000           35           41  57.363314  57.143668   \n",
       "119         0.000000           35           41  57.363314  57.143668   \n",
       "122         0.000000           32           39  59.772982  58.936270   \n",
       "123         0.000000           30           32  59.772982  59.178054   \n",
       "124         0.000000           35           41  57.890807  57.143668   \n",
       "135         0.000000           34           40  55.107337  52.831542   \n",
       "140        13.015275           34           44  56.151147  54.609355   \n",
       "189         0.000000           35           42  50.934238  50.215875   \n",
       "190         0.000000           36           42  55.697386  50.215875   \n",
       "191         0.000000           39           42  56.841631  50.215875   \n",
       "192         0.000000           33           42  52.174678  50.215875   \n",
       "193         0.000000           40           42  57.862494  50.215875   \n",
       "198         0.000000           35           42  58.483348  50.215875   \n",
       "200         0.286358           34           42  58.121724  50.215875   \n",
       "201         0.000000           30           42  52.741813  50.215875   \n",
       "218        25.862664           24           24  70.716405  68.724542   \n",
       "305        16.032730           34           38  72.484392  62.167654   \n",
       "316        13.580544           33           41  74.024399  60.885243   \n",
       "323        14.014201           28           33  73.138543  61.899024   \n",
       "325        36.755084           34           41  72.642995  63.627070   \n",
       "326        32.053479           30           31  71.738337  58.776374   \n",
       "\n",
       "         MaxGC      MinGC  MaxPlen  MinPlen  TotPrimer  \n",
       "41   66.666667  66.666667       18       18       0.50  \n",
       "51   64.705882  61.904762       21       17       0.80  \n",
       "60   66.666667  66.666667       18       18       0.50  \n",
       "118  54.545455  52.173913       23       22       1.00  \n",
       "119  54.545455  52.173913       23       22       1.00  \n",
       "122  48.000000  36.363636       33       25       0.71  \n",
       "123  44.444444  36.363636       33       27       0.71  \n",
       "124  54.545455  50.000000       24       22       1.00  \n",
       "135  70.588235  63.157895       19       17       1.00  \n",
       "140  80.000000  44.000000       25       15       1.00  \n",
       "189  61.111111  52.631579       19       18       0.80  \n",
       "190  66.666667  52.631579       19       18       0.80  \n",
       "191  63.157895  52.631579       19       19       0.80  \n",
       "192  52.631579  50.000000       20       19       0.80  \n",
       "193  72.222222  52.631579       19       18       0.80  \n",
       "198  54.166667  52.631579       24       19       0.80  \n",
       "200  56.521739  52.631579       23       19       0.80  \n",
       "201  61.111111  52.631579       19       18       0.80  \n",
       "218  47.916667  47.826087       48       46       1.00  \n",
       "305  56.521739  52.380952       42       23      20.00  \n",
       "316  54.761905  48.148148       42       27      20.00  \n",
       "323  52.380952  44.827586       42       29      20.00  \n",
       "325  51.162791  46.428571       43       28      20.00  \n",
       "326  45.833333  44.897959       49       24      20.00  \n",
       "\n",
       "[24 rows x 34 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Move outcome to the first position\n",
    "PCR_Data=PCR_Data[['outcome', 'outcome_clean','temp1', 'melting_temp2',\n",
    "       'annealing_temp2', 'extension_temp2', 'cycles2', 'temp3',\n",
    "       'band_size', 'dna_amt','t1(sec)', 'melting_t2(sec)',\n",
    "       'annealing_t3(sec)', 'extension_t4(sec)', 't5(sec)', 'Tm_p1_p3orig', 'Tm_p2_p3orig', 'Tm_p1_p3',\n",
    "       'Tm_p2_p3', 'Tm_p1_BSU', 'Tm_p2_BSU', 'Tm_dif_IDT',\n",
    "       'HPin_Tm_Max', 'Heterodimer_tm', 'Homodimer_maxTm', 'GC_ClampMin',\n",
    "       'GC_ClampMax', 'MaxTm_IDT','MinTm_IDT','MaxGC','MinGC','MaxPlen','MinPlen','TotPrimer']]\n",
    "PCR_Data[(PCR_Data['outcome']==1)&(PCR_Data['outcome_clean']==0)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Current date\n",
      "2020-11-01\n"
     ]
    }
   ],
   "source": [
    "import datetime\n",
    "now = datetime.datetime.now()\n",
    "print(\"Current date\")\n",
    "now = str(now)\n",
    "date = now.split(\" \")\n",
    "print(date[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Combined_2017-19-20_RawData2020-11-01_myeng110120_outcome_clean.csv\n"
     ]
    }
   ],
   "source": [
    "engfname = \"Combined_2017-19-20_RawData\"+date[0]+\"_myeng110120_outcome_clean.csv\"#0706 noticed error in primer feat eng\n",
    "print(engfname)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [],
   "source": [
    "cleanedF = os.path.join(\"C:/Users/jpcor/GoogleDrive/Data Science/Clauset's/PCR Analysis/CleanedData/\", engfname)  \n",
    "PCR_Data.to_csv(cleanedF,index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "\"['HP_P1' 'HP_P2'] not found in axis\"",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-63-bf66ed6207b9>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m      1\u001b[0m \u001b[1;31m#Drop HP_P1 and HP_P2\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      2\u001b[0m \u001b[0mremove\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m\"HP_P1\"\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;34m\"HP_P2\"\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 3\u001b[1;33m \u001b[0mPCR_Data\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mdrop\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mremove\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0maxis\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;36m1\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0minplace\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;32mTrue\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m      4\u001b[0m \u001b[0mPCR_Data\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mhead\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      5\u001b[0m \u001b[0mremove\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m\"Homodimer_tm1\"\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;34m\"Homodimer_tm2\"\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\Anaconda3\\lib\\site-packages\\pandas\\core\\frame.py\u001b[0m in \u001b[0;36mdrop\u001b[1;34m(self, labels, axis, index, columns, level, inplace, errors)\u001b[0m\n\u001b[0;32m   3995\u001b[0m             \u001b[0mlevel\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mlevel\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   3996\u001b[0m             \u001b[0minplace\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0minplace\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 3997\u001b[1;33m             \u001b[0merrors\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0merrors\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   3998\u001b[0m         )\n\u001b[0;32m   3999\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\Anaconda3\\lib\\site-packages\\pandas\\core\\generic.py\u001b[0m in \u001b[0;36mdrop\u001b[1;34m(self, labels, axis, index, columns, level, inplace, errors)\u001b[0m\n\u001b[0;32m   3934\u001b[0m         \u001b[1;32mfor\u001b[0m \u001b[0maxis\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mlabels\u001b[0m \u001b[1;32min\u001b[0m \u001b[0maxes\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mitems\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   3935\u001b[0m             \u001b[1;32mif\u001b[0m \u001b[0mlabels\u001b[0m \u001b[1;32mis\u001b[0m \u001b[1;32mnot\u001b[0m \u001b[1;32mNone\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 3936\u001b[1;33m                 \u001b[0mobj\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mobj\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_drop_axis\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mlabels\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mlevel\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mlevel\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0merrors\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0merrors\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   3937\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   3938\u001b[0m         \u001b[1;32mif\u001b[0m \u001b[0minplace\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\Anaconda3\\lib\\site-packages\\pandas\\core\\generic.py\u001b[0m in \u001b[0;36m_drop_axis\u001b[1;34m(self, labels, axis, level, errors)\u001b[0m\n\u001b[0;32m   3968\u001b[0m                 \u001b[0mnew_axis\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0maxis\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mdrop\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mlabels\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mlevel\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mlevel\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0merrors\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0merrors\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   3969\u001b[0m             \u001b[1;32melse\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 3970\u001b[1;33m                 \u001b[0mnew_axis\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0maxis\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mdrop\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mlabels\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0merrors\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0merrors\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   3971\u001b[0m             \u001b[0mresult\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mreindex\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m**\u001b[0m\u001b[1;33m{\u001b[0m\u001b[0maxis_name\u001b[0m\u001b[1;33m:\u001b[0m \u001b[0mnew_axis\u001b[0m\u001b[1;33m}\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   3972\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\Anaconda3\\lib\\site-packages\\pandas\\core\\indexes\\base.py\u001b[0m in \u001b[0;36mdrop\u001b[1;34m(self, labels, errors)\u001b[0m\n\u001b[0;32m   5015\u001b[0m         \u001b[1;32mif\u001b[0m \u001b[0mmask\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0many\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   5016\u001b[0m             \u001b[1;32mif\u001b[0m \u001b[0merrors\u001b[0m \u001b[1;33m!=\u001b[0m \u001b[1;34m\"ignore\"\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 5017\u001b[1;33m                 \u001b[1;32mraise\u001b[0m \u001b[0mKeyError\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34mf\"{labels[mask]} not found in axis\"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   5018\u001b[0m             \u001b[0mindexer\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mindexer\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;33m~\u001b[0m\u001b[0mmask\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   5019\u001b[0m         \u001b[1;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mdelete\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mindexer\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mKeyError\u001b[0m: \"['HP_P1' 'HP_P2'] not found in axis\""
     ]
    }
   ],
   "source": [
    "#Drop HP_P1 and HP_P2\n",
    "remove=[\"HP_P1\",\"HP_P2\"]\n",
    "PCR_Data.drop(remove,axis=1,inplace=True)\n",
    "PCR_Data.head()\n",
    "remove=[\"Homodimer_tm1\",\"Homodimer_tm2\"]\n",
    "PCR_Data.drop(remove,axis=1,inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
